// @google-cloud/recommender/build/src/v1/recommender_client.js
"use strict";
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// ** This file is automatically generated by gapic-generator-typescript. **
// ** https://github.com/googleapis/gapic-generator-typescript **
// ** All changes to this file may be overwritten. **
Object.defineProperty(exports, "__esModule", { value: true });
exports.RecommenderClient = void 0;
const jsonProtos = require("../../protos/protos.json");
const google_gax_1 = require("google-gax");
/**
 * Client JSON configuration object, loaded from
 * `src/v1/recommender_client_config.json`.
 * This file defines retry strategy and timeouts for all API methods in this library.
 */
const gapicConfig = require("./recommender_client_config.json");
const version = require('../../../package.json').version;
/**
 *  Provides insights and recommendations for cloud customers for various
 *  categories like performance optimization, cost savings, reliability, feature
 *  discovery, etc. Insights and recommendations are generated automatically
 *  based on analysis of user resources, configuration and monitoring metrics.
 * @class
 * @memberof v1
 */
class RecommenderClient {
    _terminated = false;
    _opts;
    _providedCustomServicePath;
    _gaxModule;
    _gaxGrpc;
    _protos;
    _defaults;
    _universeDomain;
    _servicePath;
    _log = google_gax_1.loggingUtils.log('recommender');
    auth;
    descriptors = {
        page: {},
        stream: {},
        longrunning: {},
        batching: {},
    };
    warn;
    innerApiCalls;
    pathTemplates;
    recommenderStub;
    /**
     * Construct an instance of RecommenderClient.
     *
     * @param {object} [options] - The configuration object.
     * The options accepted by the constructor are described in detail
     * in [this document](https://github.com/googleapis/gax-nodejs/blob/main/client-libraries.md#creating-the-client-instance).
     * The common options are:
     * @param {object} [options.credentials] - Credentials object.
     * @param {string} [options.credentials.client_email]
     * @param {string} [options.credentials.private_key]
     * @param {string} [options.email] - Account email address. Required when
     *     using a .pem or .p12 keyFilename.
     * @param {string} [options.keyFilename] - Full path to the a .json, .pem, or
     *     .p12 key downloaded from the Google Developers Console. If you provide
     *     a path to a JSON file, the projectId option below is not necessary.
     *     NOTE: .pem and .p12 require you to specify options.email as well.
     * @param {number} [options.port] - The port on which to connect to
     *     the remote host.
     * @param {string} [options.projectId] - The project ID from the Google
     *     Developer's Console, e.g. 'grape-spaceship-123'. We will also check
     *     the environment variable GCLOUD_PROJECT for your project ID. If your
     *     app is running in an environment which supports
     *     {@link https://cloud.google.com/docs/authentication/application-default-credentials Application Default Credentials},
     *     your project ID will be detected automatically.
     * @param {string} [options.apiEndpoint] - The domain name of the
     *     API remote host.
     * @param {gax.ClientConfig} [options.clientConfig] - Client configuration override.
     *     Follows the structure of {@link gapicConfig}.
     * @param {boolean} [options.fallback] - Use HTTP/1.1 REST mode.
     *     For more information, please check the
     *     {@link https://github.com/googleapis/gax-nodejs/blob/main/client-libraries.md#http11-rest-api-mode documentation}.
     * @param {gax} [gaxInstance]: loaded instance of `google-gax`. Useful if you
     *     need to avoid loading the default gRPC version and want to use the fallback
     *     HTTP implementation. Load only fallback version and pass it to the constructor:
     *     ```
     *     const gax = require('google-gax/build/src/fallback'); // avoids loading google-gax with gRPC
     *     const client = new RecommenderClient({fallback: true}, gax);
     *     ```
     */
    constructor(opts, gaxInstance) {
        // Ensure that options include all the required fields.
        const staticMembers = this.constructor;
        if (opts?.universe_domain && opts?.universeDomain && opts?.universe_domain !== opts?.universeDomain) {
            throw new Error('Please set either universe_domain or universeDomain, but not both.');
        }
        const universeDomainEnvVar = (typeof process === 'object' && typeof process.env === 'object') ? process.env['GOOGLE_CLOUD_UNIVERSE_DOMAIN'] : undefined;
        this._universeDomain = opts?.universeDomain ?? opts?.universe_domain ?? universeDomainEnvVar ?? 'googleapis.com';
        this._servicePath = 'recommender.' + this._universeDomain;
        const servicePath = opts?.servicePath || opts?.apiEndpoint || this._servicePath;
        this._providedCustomServicePath = !!(opts?.servicePath || opts?.apiEndpoint);
        const port = opts?.port || staticMembers.port;
        const clientConfig = opts?.clientConfig ?? {};
        const fallback = opts?.fallback ?? (typeof window !== 'undefined' && typeof window?.fetch === 'function');
        opts = Object.assign({ servicePath, port, clientConfig, fallback }, opts);
        // Request numeric enum values if REST transport is used.
        opts.numericEnums = true;
        // If scopes are unset in options and we're connecting to a non-default endpoint, set scopes just in case.
        if (servicePath !== this._servicePath && !('scopes' in opts)) {
            opts['scopes'] = staticMembers.scopes;
        }
        // Load google-gax module synchronously if needed
        if (!gaxInstance) {
            gaxInstance = require('google-gax');
        }
        // Choose either gRPC or proto-over-HTTP implementation of google-gax.
        this._gaxModule = opts.fallback ? gaxInstance.fallback : gaxInstance;
        // Create a `gaxGrpc` object, with any grpc-specific options sent to the client.
        this._gaxGrpc = new this._gaxModule.GrpcClient(opts);
        // Save options to use in initialize() method.
        this._opts = opts;
        // Save the auth object to the client, for use by other methods.
        this.auth = this._gaxGrpc.auth;
        // Set useJWTAccessWithScope on the auth object.
        this.auth.useJWTAccessWithScope = true;
        // Set defaultServicePath on the auth object.
        this.auth.defaultServicePath = this._servicePath;
        // Set the default scopes in auth client if needed.
        if (servicePath === this._servicePath) {
            this.auth.defaultScopes = staticMembers.scopes;
        }
        // Determine the client header string.
        const clientHeader = [
            `gax/${this._gaxModule.version}`,
            `gapic/${version}`,
        ];
        if (typeof process === 'object' && 'versions' in process) {
            clientHeader.push(`gl-node/${process.versions.node}`);
        }
        else {
            clientHeader.push(`gl-web/${this._gaxModule.version}`);
        }
        if (!opts.fallback) {
            clientHeader.push(`grpc/${this._gaxGrpc.grpcVersion}`);
        }
        else {
            clientHeader.push(`rest/${this._gaxGrpc.grpcVersion}`);
        }
        if (opts.libName && opts.libVersion) {
            clientHeader.push(`${opts.libName}/${opts.libVersion}`);
        }
        // Load the applicable protos.
        this._protos = this._gaxGrpc.loadProtoJSON(jsonProtos);
        // This API contains "path templates"; forward-slash-separated
        // identifiers to uniquely identify resources within the API.
        // Create useful helper objects for these.
        this.pathTemplates = {
            billingAccountLocationInsightTypePathTemplate: new this._gaxModule.PathTemplate('billingAccounts/{billing_account}/locations/{location}/insightTypes/{insight_type}'),
            billingAccountLocationInsightTypeConfigPathTemplate: new this._gaxModule.PathTemplate('billingAccounts/{billing_account}/locations/{location}/insightTypes/{insight_type}/config'),
            billingAccountLocationInsightTypeInsightPathTemplate: new this._gaxModule.PathTemplate('billingAccounts/{billing_account}/locations/{location}/insightTypes/{insight_type}/insights/{insight}'),
            billingAccountLocationRecommenderPathTemplate: new this._gaxModule.PathTemplate('billingAccounts/{billing_account}/locations/{location}/recommenders/{recommender}'),
            billingAccountLocationRecommenderConfigPathTemplate: new this._gaxModule.PathTemplate('billingAccounts/{billing_account}/locations/{location}/recommenders/{recommender}/config'),
            billingAccountLocationRecommenderRecommendationPathTemplate: new this._gaxModule.PathTemplate('billingAccounts/{billing_account}/locations/{location}/recommenders/{recommender}/recommendations/{recommendation}'),
            folderLocationInsightTypePathTemplate: new this._gaxModule.PathTemplate('folders/{folder}/locations/{location}/insightTypes/{insight_type}'),
            folderLocationInsightTypeInsightPathTemplate: new this._gaxModule.PathTemplate('folders/{folder}/locations/{location}/insightTypes/{insight_type}/insights/{insight}'),
            folderLocationRecommenderPathTemplate: new this._gaxModule.PathTemplate('folders/{folder}/locations/{location}/recommenders/{recommender}'),
            folderLocationRecommenderRecommendationPathTemplate: new this._gaxModule.PathTemplate('folders/{folder}/locations/{location}/recommenders/{recommender}/recommendations/{recommendation}'),
            organizationLocationInsightTypePathTemplate: new this._gaxModule.PathTemplate('organizations/{organization}/locations/{location}/insightTypes/{insight_type}'),
            organizationLocationInsightTypeConfigPathTemplate: new this._gaxModule.PathTemplate('organizations/{organization}/locations/{location}/insightTypes/{insight_type}/config'),
            organizationLocationInsightTypeInsightPathTemplate: new this._gaxModule.PathTemplate('organizations/{organization}/locations/{location}/insightTypes/{insight_type}/insights/{insight}'),
            organizationLocationRecommenderPathTemplate: new this._gaxModule.PathTemplate('organizations/{organization}/locations/{location}/recommenders/{recommender}'),
            organizationLocationRecommenderConfigPathTemplate: new this._gaxModule.PathTemplate('organizations/{organization}/locations/{location}/recommenders/{recommender}/config'),
            organizationLocationRecommenderRecommendationPathTemplate: new this._gaxModule.PathTemplate('organizations/{organization}/locations/{location}/recommenders/{recommender}/recommendations/{recommendation}'),
            projectLocationInsightTypePathTemplate: new this._gaxModule.PathTemplate('projects/{project}/locations/{location}/insightTypes/{insight_type}'),
            projectLocationInsightTypeConfigPathTemplate: new this._gaxModule.PathTemplate('projects/{project}/locations/{location}/insightTypes/{insight_type}/config'),
            projectLocationInsightTypeInsightPathTemplate: new this._gaxModule.PathTemplate('projects/{project}/locations/{location}/insightTypes/{insight_type}/insights/{insight}'),
            projectLocationRecommenderPathTemplate: new this._gaxModule.PathTemplate('projects/{project}/locations/{location}/recommenders/{recommender}'),
            projectLocationRecommenderConfigPathTemplate: new this._gaxModule.PathTemplate('projects/{project}/locations/{location}/recommenders/{recommender}/config'),
            projectLocationRecommenderRecommendationPathTemplate: new this._gaxModule.PathTemplate('projects/{project}/locations/{location}/recommenders/{recommender}/recommendations/{recommendation}'),
        };
        // Some of the methods on this service return "paged" results,
        // (e.g. 50 results at a time, with tokens to get subsequent
        // pages). Denote the keys used for pagination and results.
        this.descriptors.page = {
            listInsights: new this._gaxModule.PageDescriptor('pageToken', 'nextPageToken', 'insights'),
            listRecommendations: new this._gaxModule.PageDescriptor('pageToken', 'nextPageToken', 'recommendations')
        };
        // Put together the default options sent with requests.
        this._defaults = this._gaxGrpc.constructSettings('google.cloud.recommender.v1.Recommender', gapicConfig, opts.clientConfig || {}, { 'x-goog-api-client': clientHeader.join(' ') });
        // Set up a dictionary of "inner API calls"; the core implementation
        // of calling the API is handled in `google-gax`, with this code
        // merely providing the destination and request information.
        this.innerApiCalls = {};
        // Add a warn function to the client constructor so it can be easily tested.
        this.warn = this._gaxModule.warn;
    }
    /**
     * Initialize the client.
     * Performs asynchronous operations (such as authentication) and prepares the client.
     * This function will be called automatically when any class method is called for the
     * first time, but if you need to initialize it before calling an actual method,
     * feel free to call initialize() directly.
     *
     * You can await on this method if you want to make sure the client is initialized.
     *
     * @returns {Promise} A promise that resolves to an authenticated service stub.
     */
    initialize() {
        // If the client stub promise is already initialized, return immediately.
        if (this.recommenderStub) {
            return this.recommenderStub;
        }
        // Put together the "service stub" for
        // google.cloud.recommender.v1.Recommender.
        this.recommenderStub = this._gaxGrpc.createStub(this._opts.fallback ?
            this._protos.lookupService('google.cloud.recommender.v1.Recommender') :
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            this._protos.google.cloud.recommender.v1.Recommender, this._opts, this._providedCustomServicePath);
        // Iterate over each of the methods that the service provides
        // and create an API call method for each.
        const recommenderStubMethods = ['listInsights', 'getInsight', 'markInsightAccepted', 'listRecommendations', 'getRecommendation', 'markRecommendationDismissed', 'markRecommendationClaimed', 'markRecommendationSucceeded', 'markRecommendationFailed', 'getRecommenderConfig', 'updateRecommenderConfig', 'getInsightTypeConfig', 'updateInsightTypeConfig'];
        for (const methodName of recommenderStubMethods) {
            const callPromise = this.recommenderStub.then(stub => (...args) => {
                if (this._terminated) {
                    return Promise.reject('The client has already been closed.');
                }
                const func = stub[methodName];
                return func.apply(stub, args);
            }, (err) => () => {
                throw err;
            });
            const descriptor = this.descriptors.page[methodName] ||
                undefined;
            const apiCall = this._gaxModule.createApiCall(callPromise, this._defaults[methodName], descriptor, this._opts.fallback);
            this.innerApiCalls[methodName] = apiCall;
        }
        return this.recommenderStub;
    }
    /**
     * The DNS address for this API service.
     * @deprecated Use the apiEndpoint method of the client instance.
     * @returns {string} The DNS address for this service.
     */
    static get servicePath() {
        if (typeof process === 'object' && typeof process.emitWarning === 'function') {
            process.emitWarning('Static servicePath is deprecated, please use the instance method instead.', 'DeprecationWarning');
        }
        return 'recommender.googleapis.com';
    }
    /**
     * The DNS address for this API service - same as servicePath.
     * @deprecated Use the apiEndpoint method of the client instance.
     * @returns {string} The DNS address for this service.
     */
    static get apiEndpoint() {
        if (typeof process === 'object' && typeof process.emitWarning === 'function') {
            process.emitWarning('Static apiEndpoint is deprecated, please use the instance method instead.', 'DeprecationWarning');
        }
        return 'recommender.googleapis.com';
    }
    /**
     * The DNS address for this API service.
     * @returns {string} The DNS address for this service.
     */
    get apiEndpoint() {
        return this._servicePath;
    }
    get universeDomain() {
        return this._universeDomain;
    }
    /**
     * The port for this API service.
     * @returns {number} The default port for this service.
     */
    static get port() {
        return 443;
    }
    /**
     * The scopes needed to make gRPC calls for every method defined
     * in this service.
     * @returns {string[]} List of default scopes.
     */
    static get scopes() {
        return [
            'https://www.googleapis.com/auth/cloud-platform'
        ];
    }
    /**
     * Return the project ID used by this class.
     * @returns {Promise} A promise that resolves to string containing the project ID.
     */
    getProjectId(callback) {
        if (callback) {
            this.auth.getProjectId(callback);
            return;
        }
        return this.auth.getProjectId();
    }
    getInsight(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'name': request.name ?? '',
        });
        this.initialize().catch(err => { throw err; });
        this._log.info('getInsight request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('getInsight response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls.getInsight(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('getInsight response %j', response);
            return [response, options, rawResponse];
        }).catch((error) => {
            if (error && 'statusDetails' in error && error.statusDetails instanceof Array) {
                const protos = this._gaxModule.protobuf.Root.fromJSON(jsonProtos);
                error.statusDetails = (0, google_gax_1.decodeAnyProtosInArray)(error.statusDetails, protos);
            }
            throw error;
        });
    }
    markInsightAccepted(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'name': request.name ?? '',
        });
        this.initialize().catch(err => { throw err; });
        this._log.info('markInsightAccepted request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('markInsightAccepted response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls.markInsightAccepted(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('markInsightAccepted response %j', response);
            return [response, options, rawResponse];
        }).catch((error) => {
            if (error && 'statusDetails' in error && error.statusDetails instanceof Array) {
                const protos = this._gaxModule.protobuf.Root.fromJSON(jsonProtos);
                error.statusDetails = (0, google_gax_1.decodeAnyProtosInArray)(error.statusDetails, protos);
            }
            throw error;
        });
    }
    getRecommendation(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'name': request.name ?? '',
        });
        this.initialize().catch(err => { throw err; });
        this._log.info('getRecommendation request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('getRecommendation response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls.getRecommendation(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('getRecommendation response %j', response);
            return [response, options, rawResponse];
        }).catch((error) => {
            if (error && 'statusDetails' in error && error.statusDetails instanceof Array) {
                const protos = this._gaxModule.protobuf.Root.fromJSON(jsonProtos);
                error.statusDetails = (0, google_gax_1.decodeAnyProtosInArray)(error.statusDetails, protos);
            }
            throw error;
        });
    }
    markRecommendationDismissed(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'name': request.name ?? '',
        });
        this.initialize().catch(err => { throw err; });
        this._log.info('markRecommendationDismissed request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('markRecommendationDismissed response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls.markRecommendationDismissed(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('markRecommendationDismissed response %j', response);
            return [response, options, rawResponse];
        }).catch((error) => {
            if (error && 'statusDetails' in error && error.statusDetails instanceof Array) {
                const protos = this._gaxModule.protobuf.Root.fromJSON(jsonProtos);
                error.statusDetails = (0, google_gax_1.decodeAnyProtosInArray)(error.statusDetails, protos);
            }
            throw error;
        });
    }
    markRecommendationClaimed(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'name': request.name ?? '',
        });
        this.initialize().catch(err => { throw err; });
        this._log.info('markRecommendationClaimed request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('markRecommendationClaimed response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls.markRecommendationClaimed(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('markRecommendationClaimed response %j', response);
            return [response, options, rawResponse];
        }).catch((error) => {
            if (error && 'statusDetails' in error && error.statusDetails instanceof Array) {
                const protos = this._gaxModule.protobuf.Root.fromJSON(jsonProtos);
                error.statusDetails = (0, google_gax_1.decodeAnyProtosInArray)(error.statusDetails, protos);
            }
            throw error;
        });
    }
    markRecommendationSucceeded(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'name': request.name ?? '',
        });
        this.initialize().catch(err => { throw err; });
        this._log.info('markRecommendationSucceeded request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('markRecommendationSucceeded response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls.markRecommendationSucceeded(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('markRecommendationSucceeded response %j', response);
            return [response, options, rawResponse];
        }).catch((error) => {
            if (error && 'statusDetails' in error && error.statusDetails instanceof Array) {
                const protos = this._gaxModule.protobuf.Root.fromJSON(jsonProtos);
                error.statusDetails = (0, google_gax_1.decodeAnyProtosInArray)(error.statusDetails, protos);
            }
            throw error;
        });
    }
    markRecommendationFailed(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'name': request.name ?? '',
        });
        this.initialize().catch(err => { throw err; });
        this._log.info('markRecommendationFailed request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('markRecommendationFailed response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls.markRecommendationFailed(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('markRecommendationFailed response %j', response);
            return [response, options, rawResponse];
        }).catch((error) => {
            if (error && 'statusDetails' in error && error.statusDetails instanceof Array) {
                const protos = this._gaxModule.protobuf.Root.fromJSON(jsonProtos);
                error.statusDetails = (0, google_gax_1.decodeAnyProtosInArray)(error.statusDetails, protos);
            }
            throw error;
        });
    }
    getRecommenderConfig(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'name': request.name ?? '',
        });
        this.initialize().catch(err => { throw err; });
        this._log.info('getRecommenderConfig request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('getRecommenderConfig response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls.getRecommenderConfig(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('getRecommenderConfig response %j', response);
            return [response, options, rawResponse];
        }).catch((error) => {
            if (error && 'statusDetails' in error && error.statusDetails instanceof Array) {
                const protos = this._gaxModule.protobuf.Root.fromJSON(jsonProtos);
                error.statusDetails = (0, google_gax_1.decodeAnyProtosInArray)(error.statusDetails, protos);
            }
            throw error;
        });
    }
    updateRecommenderConfig(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'recommender_config.name': request.recommenderConfig.name ?? '',
        });
        this.initialize().catch(err => { throw err; });
        this._log.info('updateRecommenderConfig request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('updateRecommenderConfig response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls.updateRecommenderConfig(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('updateRecommenderConfig response %j', response);
            return [response, options, rawResponse];
        }).catch((error) => {
            if (error && 'statusDetails' in error && error.statusDetails instanceof Array) {
                const protos = this._gaxModule.protobuf.Root.fromJSON(jsonProtos);
                error.statusDetails = (0, google_gax_1.decodeAnyProtosInArray)(error.statusDetails, protos);
            }
            throw error;
        });
    }
    getInsightTypeConfig(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'name': request.name ?? '',
        });
        this.initialize().catch(err => { throw err; });
        this._log.info('getInsightTypeConfig request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('getInsightTypeConfig response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls.getInsightTypeConfig(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('getInsightTypeConfig response %j', response);
            return [response, options, rawResponse];
        }).catch((error) => {
            if (error && 'statusDetails' in error && error.statusDetails instanceof Array) {
                const protos = this._gaxModule.protobuf.Root.fromJSON(jsonProtos);
                error.statusDetails = (0, google_gax_1.decodeAnyProtosInArray)(error.statusDetails, protos);
            }
            throw error;
        });
    }
    updateInsightTypeConfig(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'insight_type_config.name': request.insightTypeConfig.name ?? '',
        });
        this.initialize().catch(err => { throw err; });
        this._log.info('updateInsightTypeConfig request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('updateInsightTypeConfig response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls.updateInsightTypeConfig(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('updateInsightTypeConfig response %j', response);
            return [response, options, rawResponse];
        }).catch((error) => {
            if (error && 'statusDetails' in error && error.statusDetails instanceof Array) {
                const protos = this._gaxModule.protobuf.Root.fromJSON(jsonProtos);
                error.statusDetails = (0, google_gax_1.decodeAnyProtosInArray)(error.statusDetails, protos);
            }
            throw error;
        });
    }
    listInsights(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'parent': request.parent ?? '',
        });
        this.initialize().catch(err => { throw err; });
        const wrappedCallback = callback
            ? (error, values, nextPageRequest, rawResponse) => {
                this._log.info('listInsights values %j', values);
                callback(error, values, nextPageRequest, rawResponse); // We verified callback above.
            }
            : undefined;
        this._log.info('listInsights request %j', request);
        return this.innerApiCalls
            .listInsights(request, options, wrappedCallback)
            ?.then(([response, input, output]) => {
            this._log.info('listInsights values %j', response);
            return [response, input, output];
        });
    }
    /**
     * Equivalent to `listInsights`, but returns a NodeJS Stream object.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.parent
     *   Required. The container resource on which to execute the request.
     *   Acceptable formats:
     *
     *   * `projects/[PROJECT_NUMBER]/locations/[LOCATION]/insightTypes/[INSIGHT_TYPE_ID]`
     *
     *   * `projects/[PROJECT_ID]/locations/[LOCATION]/insightTypes/[INSIGHT_TYPE_ID]`
     *
     *   * `billingAccounts/[BILLING_ACCOUNT_ID]/locations/[LOCATION]/insightTypes/[INSIGHT_TYPE_ID]`
     *
     *   * `folders/[FOLDER_ID]/locations/[LOCATION]/insightTypes/[INSIGHT_TYPE_ID]`
     *
     *   * `organizations/[ORGANIZATION_ID]/locations/[LOCATION]/insightTypes/[INSIGHT_TYPE_ID]`
     *
     *   LOCATION here refers to GCP Locations:
     *   https://cloud.google.com/about/locations/
     *   INSIGHT_TYPE_ID refers to supported insight types:
     *   https://cloud.google.com/recommender/docs/insights/insight-types.
     * @param {number} [request.pageSize]
     *   Optional. The maximum number of results to return from this request.
     *   Non-positive values are ignored. If not specified, the server will
     *   determine the number of results to return.
     * @param {string} [request.pageToken]
     *   Optional. If present, retrieves the next batch of results from the
     *   preceding call to this method. `page_token` must be the value of
     *   `next_page_token` from the previous response. The values of other method
     *   parameters must be identical to those in the previous call.
     * @param {string} [request.filter]
     *   Optional. Filter expression to restrict the insights returned. Supported
     *   filter fields:
     *
     *   * `stateInfo.state`
     *
     *   * `insightSubtype`
     *
     *   * `severity`
     *
     *   * `targetResources`
     *
     *   Examples:
     *
     *   * `stateInfo.state = ACTIVE OR stateInfo.state = DISMISSED`
     *
     *   * `insightSubtype = PERMISSIONS_USAGE`
     *
     *   * `severity = CRITICAL OR severity = HIGH`
     *
     *   * `targetResources :
     *   //compute.googleapis.com/projects/1234/zones/us-central1-a/instances/instance-1`
     *
     *   * `stateInfo.state = ACTIVE AND (severity = CRITICAL OR severity = HIGH)`
     *
     *   The max allowed filter length is 500 characters.
     *
     *   (These expressions are based on the filter language described at
     *   https://google.aip.dev/160)
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Stream}
     *   An object stream which emits an object representing {@link protos.google.cloud.recommender.v1.Insight|Insight} on 'data' event.
     *   The client library will perform auto-pagination by default: it will call the API as many
     *   times as needed. Note that it can affect your quota.
     *   We recommend using `listInsightsAsync()`
     *   method described below for async iteration which you can stop as needed.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     */
    listInsightsStream(request, options) {
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'parent': request.parent ?? '',
        });
        const defaultCallSettings = this._defaults['listInsights'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize().catch(err => { throw err; });
        this._log.info('listInsights stream %j', request);
        return this.descriptors.page.listInsights.createStream(this.innerApiCalls.listInsights, request, callSettings);
    }
    /**
     * Equivalent to `listInsights`, but returns an iterable object.
     *
     * `for`-`await`-`of` syntax is used with the iterable to get response elements on-demand.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.parent
     *   Required. The container resource on which to execute the request.
     *   Acceptable formats:
     *
     *   * `projects/[PROJECT_NUMBER]/locations/[LOCATION]/insightTypes/[INSIGHT_TYPE_ID]`
     *
     *   * `projects/[PROJECT_ID]/locations/[LOCATION]/insightTypes/[INSIGHT_TYPE_ID]`
     *
     *   * `billingAccounts/[BILLING_ACCOUNT_ID]/locations/[LOCATION]/insightTypes/[INSIGHT_TYPE_ID]`
     *
     *   * `folders/[FOLDER_ID]/locations/[LOCATION]/insightTypes/[INSIGHT_TYPE_ID]`
     *
     *   * `organizations/[ORGANIZATION_ID]/locations/[LOCATION]/insightTypes/[INSIGHT_TYPE_ID]`
     *
     *   LOCATION here refers to GCP Locations:
     *   https://cloud.google.com/about/locations/
     *   INSIGHT_TYPE_ID refers to supported insight types:
     *   https://cloud.google.com/recommender/docs/insights/insight-types.
     * @param {number} [request.pageSize]
     *   Optional. The maximum number of results to return from this request.
     *   Non-positive values are ignored. If not specified, the server will
     *   determine the number of results to return.
     * @param {string} [request.pageToken]
     *   Optional. If present, retrieves the next batch of results from the
     *   preceding call to this method. `page_token` must be the value of
     *   `next_page_token` from the previous response. The values of other method
     *   parameters must be identical to those in the previous call.
     * @param {string} [request.filter]
     *   Optional. Filter expression to restrict the insights returned. Supported
     *   filter fields:
     *
     *   * `stateInfo.state`
     *
     *   * `insightSubtype`
     *
     *   * `severity`
     *
     *   * `targetResources`
     *
     *   Examples:
     *
     *   * `stateInfo.state = ACTIVE OR stateInfo.state = DISMISSED`
     *
     *   * `insightSubtype = PERMISSIONS_USAGE`
     *
     *   * `severity = CRITICAL OR severity = HIGH`
     *
     *   * `targetResources :
     *   //compute.googleapis.com/projects/1234/zones/us-central1-a/instances/instance-1`
     *
     *   * `stateInfo.state = ACTIVE AND (severity = CRITICAL OR severity = HIGH)`
     *
     *   The max allowed filter length is 500 characters.
     *
     *   (These expressions are based on the filter language described at
     *   https://google.aip.dev/160)
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Object}
     *   An iterable Object that allows {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols | async iteration }.
     *   When you iterate the returned iterable, each element will be an object representing
     *   {@link protos.google.cloud.recommender.v1.Insight|Insight}. The API will be called under the hood as needed, once per the page,
     *   so you can stop the iteration when you don't need more results.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     * @example <caption>include:samples/generated/v1/recommender.list_insights.js</caption>
     * region_tag:recommender_v1_generated_Recommender_ListInsights_async
     */
    listInsightsAsync(request, options) {
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'parent': request.parent ?? '',
        });
        const defaultCallSettings = this._defaults['listInsights'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize().catch(err => { throw err; });
        this._log.info('listInsights iterate %j', request);
        return this.descriptors.page.listInsights.asyncIterate(this.innerApiCalls['listInsights'], request, callSettings);
    }
    listRecommendations(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'parent': request.parent ?? '',
        });
        this.initialize().catch(err => { throw err; });
        const wrappedCallback = callback
            ? (error, values, nextPageRequest, rawResponse) => {
                this._log.info('listRecommendations values %j', values);
                callback(error, values, nextPageRequest, rawResponse); // We verified callback above.
            }
            : undefined;
        this._log.info('listRecommendations request %j', request);
        return this.innerApiCalls
            .listRecommendations(request, options, wrappedCallback)
            ?.then(([response, input, output]) => {
            this._log.info('listRecommendations values %j', response);
            return [response, input, output];
        });
    }
    /**
     * Equivalent to `listRecommendations`, but returns a NodeJS Stream object.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.parent
     *   Required. The container resource on which to execute the request.
     *   Acceptable formats:
     *
     *   * `projects/[PROJECT_NUMBER]/locations/[LOCATION]/recommenders/[RECOMMENDER_ID]`
     *
     *   * `projects/[PROJECT_ID]/locations/[LOCATION]/recommenders/[RECOMMENDER_ID]`
     *
     *   * `billingAccounts/[BILLING_ACCOUNT_ID]/locations/[LOCATION]/recommenders/[RECOMMENDER_ID]`
     *
     *   * `folders/[FOLDER_ID]/locations/[LOCATION]/recommenders/[RECOMMENDER_ID]`
     *
     *   * `organizations/[ORGANIZATION_ID]/locations/[LOCATION]/recommenders/[RECOMMENDER_ID]`
     *
     *   LOCATION here refers to GCP Locations:
     *   https://cloud.google.com/about/locations/
     *   RECOMMENDER_ID refers to supported recommenders:
     *   https://cloud.google.com/recommender/docs/recommenders.
     * @param {number} [request.pageSize]
     *   Optional. The maximum number of results to return from this request.
     *   Non-positive values are ignored. If not specified, the server will
     *   determine the number of results to return.
     * @param {string} [request.pageToken]
     *   Optional. If present, retrieves the next batch of results from the
     *   preceding call to this method. `page_token` must be the value of
     *   `next_page_token` from the previous response. The values of other method
     *   parameters must be identical to those in the previous call.
     * @param {string} request.filter
     *   Filter expression to restrict the recommendations returned. Supported
     *   filter fields:
     *
     *   * `state_info.state`
     *
     *   * `recommenderSubtype`
     *
     *   * `priority`
     *
     *   * `targetResources`
     *
     *   Examples:
     *
     *   * `stateInfo.state = ACTIVE OR stateInfo.state = DISMISSED`
     *
     *   * `recommenderSubtype = REMOVE_ROLE OR recommenderSubtype = REPLACE_ROLE`
     *
     *   * `priority = P1 OR priority = P2`
     *
     *   * `targetResources :
     *   //compute.googleapis.com/projects/1234/zones/us-central1-a/instances/instance-1`
     *
     *   * `stateInfo.state = ACTIVE AND (priority = P1 OR priority = P2)`
     *
     *   The max allowed filter length is 500 characters.
     *
     *   (These expressions are based on the filter language described at
     *   https://google.aip.dev/160)
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Stream}
     *   An object stream which emits an object representing {@link protos.google.cloud.recommender.v1.Recommendation|Recommendation} on 'data' event.
     *   The client library will perform auto-pagination by default: it will call the API as many
     *   times as needed. Note that it can affect your quota.
     *   We recommend using `listRecommendationsAsync()`
     *   method described below for async iteration which you can stop as needed.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     */
    listRecommendationsStream(request, options) {
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'parent': request.parent ?? '',
        });
        const defaultCallSettings = this._defaults['listRecommendations'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize().catch(err => { throw err; });
        this._log.info('listRecommendations stream %j', request);
        return this.descriptors.page.listRecommendations.createStream(this.innerApiCalls.listRecommendations, request, callSettings);
    }
    /**
     * Equivalent to `listRecommendations`, but returns an iterable object.
     *
     * `for`-`await`-`of` syntax is used with the iterable to get response elements on-demand.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.parent
     *   Required. The container resource on which to execute the request.
     *   Acceptable formats:
     *
     *   * `projects/[PROJECT_NUMBER]/locations/[LOCATION]/recommenders/[RECOMMENDER_ID]`
     *
     *   * `projects/[PROJECT_ID]/locations/[LOCATION]/recommenders/[RECOMMENDER_ID]`
     *
     *   * `billingAccounts/[BILLING_ACCOUNT_ID]/locations/[LOCATION]/recommenders/[RECOMMENDER_ID]`
     *
     *   * `folders/[FOLDER_ID]/locations/[LOCATION]/recommenders/[RECOMMENDER_ID]`
     *
     *   * `organizations/[ORGANIZATION_ID]/locations/[LOCATION]/recommenders/[RECOMMENDER_ID]`
     *
     *   LOCATION here refers to GCP Locations:
     *   https://cloud.google.com/about/locations/
     *   RECOMMENDER_ID refers to supported recommenders:
     *   https://cloud.google.com/recommender/docs/recommenders.
     * @param {number} [request.pageSize]
     *   Optional. The maximum number of results to return from this request.
     *   Non-positive values are ignored. If not specified, the server will
     *   determine the number of results to return.
     * @param {string} [request.pageToken]
     *   Optional. If present, retrieves the next batch of results from the
     *   preceding call to this method. `page_token` must be the value of
     *   `next_page_token` from the previous response. The values of other method
     *   parameters must be identical to those in the previous call.
     * @param {string} request.filter
     *   Filter expression to restrict the recommendations returned. Supported
     *   filter fields:
     *
     *   * `state_info.state`
     *
     *   * `recommenderSubtype`
     *
     *   * `priority`
     *
     *   * `targetResources`
     *
     *   Examples:
     *
     *   * `stateInfo.state = ACTIVE OR stateInfo.state = DISMISSED`
     *
     *   * `recommenderSubtype = REMOVE_ROLE OR recommenderSubtype = REPLACE_ROLE`
     *
     *   * `priority = P1 OR priority = P2`
     *
     *   * `targetResources :
     *   //compute.googleapis.com/projects/1234/zones/us-central1-a/instances/instance-1`
     *
     *   * `stateInfo.state = ACTIVE AND (priority = P1 OR priority = P2)`
     *
     *   The max allowed filter length is 500 characters.
     *
     *   (These expressions are based on the filter language described at
     *   https://google.aip.dev/160)
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Object}
     *   An iterable Object that allows {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols | async iteration }.
     *   When you iterate the returned iterable, each element will be an object representing
     *   {@link protos.google.cloud.recommender.v1.Recommendation|Recommendation}. The API will be called under the hood as needed, once per the page,
     *   so you can stop the iteration when you don't need more results.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     * @example <caption>include:samples/generated/v1/recommender.list_recommendations.js</caption>
     * region_tag:recommender_v1_generated_Recommender_ListRecommendations_async
     */
    listRecommendationsAsync(request, options) {
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'parent': request.parent ?? '',
        });
        const defaultCallSettings = this._defaults['listRecommendations'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize().catch(err => { throw err; });
        this._log.info('listRecommendations iterate %j', request);
        return this.descriptors.page.listRecommendations.asyncIterate(this.innerApiCalls['listRecommendations'], request, callSettings);
    }
    // --------------------
    // -- Path templates --
    // --------------------
    /**
     * Return a fully-qualified billingAccountLocationInsightType resource name string.
     *
     * @param {string} billing_account
     * @param {string} location
     * @param {string} insight_type
     * @returns {string} Resource name string.
     */
    billingAccountLocationInsightTypePath(billingAccount, location, insightType) {
        return this.pathTemplates.billingAccountLocationInsightTypePathTemplate.render({
            billing_account: billingAccount,
            location: location,
            insight_type: insightType,
        });
    }
    /**
     * Parse the billing_account from BillingAccountLocationInsightType resource.
     *
     * @param {string} billingAccountLocationInsightTypeName
     *   A fully-qualified path representing billing_account_location_insight_type resource.
     * @returns {string} A string representing the billing_account.
     */
    matchBillingAccountFromBillingAccountLocationInsightTypeName(billingAccountLocationInsightTypeName) {
        return this.pathTemplates.billingAccountLocationInsightTypePathTemplate.match(billingAccountLocationInsightTypeName).billing_account;
    }
    /**
     * Parse the location from BillingAccountLocationInsightType resource.
     *
     * @param {string} billingAccountLocationInsightTypeName
     *   A fully-qualified path representing billing_account_location_insight_type resource.
     * @returns {string} A string representing the location.
     */
    matchLocationFromBillingAccountLocationInsightTypeName(billingAccountLocationInsightTypeName) {
        return this.pathTemplates.billingAccountLocationInsightTypePathTemplate.match(billingAccountLocationInsightTypeName).location;
    }
    /**
     * Parse the insight_type from BillingAccountLocationInsightType resource.
     *
     * @param {string} billingAccountLocationInsightTypeName
     *   A fully-qualified path representing billing_account_location_insight_type resource.
     * @returns {string} A string representing the insight_type.
     */
    matchInsightTypeFromBillingAccountLocationInsightTypeName(billingAccountLocationInsightTypeName) {
        return this.pathTemplates.billingAccountLocationInsightTypePathTemplate.match(billingAccountLocationInsightTypeName).insight_type;
    }
    /**
     * Return a fully-qualified billingAccountLocationInsightTypeConfig resource name string.
     *
     * @param {string} billing_account
     * @param {string} location
     * @param {string} insight_type
     * @returns {string} Resource name string.
     */
    billingAccountLocationInsightTypeConfigPath(billingAccount, location, insightType) {
        return this.pathTemplates.billingAccountLocationInsightTypeConfigPathTemplate.render({
            billing_account: billingAccount,
            location: location,
            insight_type: insightType,
        });
    }
    /**
     * Parse the billing_account from BillingAccountLocationInsightTypeConfig resource.
     *
     * @param {string} billingAccountLocationInsightTypeConfigName
     *   A fully-qualified path representing billing_account_location_insight_type_config resource.
     * @returns {string} A string representing the billing_account.
     */
    matchBillingAccountFromBillingAccountLocationInsightTypeConfigName(billingAccountLocationInsightTypeConfigName) {
        return this.pathTemplates.billingAccountLocationInsightTypeConfigPathTemplate.match(billingAccountLocationInsightTypeConfigName).billing_account;
    }
    /**
     * Parse the location from BillingAccountLocationInsightTypeConfig resource.
     *
     * @param {string} billingAccountLocationInsightTypeConfigName
     *   A fully-qualified path representing billing_account_location_insight_type_config resource.
     * @returns {string} A string representing the location.
     */
    matchLocationFromBillingAccountLocationInsightTypeConfigName(billingAccountLocationInsightTypeConfigName) {
        return this.pathTemplates.billingAccountLocationInsightTypeConfigPathTemplate.match(billingAccountLocationInsightTypeConfigName).location;
    }
    /**
     * Parse the insight_type from BillingAccountLocationInsightTypeConfig resource.
     *
     * @param {string} billingAccountLocationInsightTypeConfigName
     *   A fully-qualified path representing billing_account_location_insight_type_config resource.
     * @returns {string} A string representing the insight_type.
     */
    matchInsightTypeFromBillingAccountLocationInsightTypeConfigName(billingAccountLocationInsightTypeConfigName) {
        return this.pathTemplates.billingAccountLocationInsightTypeConfigPathTemplate.match(billingAccountLocationInsightTypeConfigName).insight_type;
    }
    /**
     * Return a fully-qualified billingAccountLocationInsightTypeInsight resource name string.
     *
     * @param {string} billing_account
     * @param {string} location
     * @param {string} insight_type
     * @param {string} insight
     * @returns {string} Resource name string.
     */
    billingAccountLocationInsightTypeInsightPath(billingAccount, location, insightType, insight) {
        return this.pathTemplates.billingAccountLocationInsightTypeInsightPathTemplate.render({
            billing_account: billingAccount,
            location: location,
            insight_type: insightType,
            insight: insight,
        });
    }
    /**
     * Parse the billing_account from BillingAccountLocationInsightTypeInsight resource.
     *
     * @param {string} billingAccountLocationInsightTypeInsightName
     *   A fully-qualified path representing billing_account_location_insight_type_insight resource.
     * @returns {string} A string representing the billing_account.
     */
    matchBillingAccountFromBillingAccountLocationInsightTypeInsightName(billingAccountLocationInsightTypeInsightName) {
        return this.pathTemplates.billingAccountLocationInsightTypeInsightPathTemplate.match(billingAccountLocationInsightTypeInsightName).billing_account;
    }
    /**
     * Parse the location from BillingAccountLocationInsightTypeInsight resource.
     *
     * @param {string} billingAccountLocationInsightTypeInsightName
     *   A fully-qualified path representing billing_account_location_insight_type_insight resource.
     * @returns {string} A string representing the location.
     */
    matchLocationFromBillingAccountLocationInsightTypeInsightName(billingAccountLocationInsightTypeInsightName) {
        return this.pathTemplates.billingAccountLocationInsightTypeInsightPathTemplate.match(billingAccountLocationInsightTypeInsightName).location;
    }
    /**
     * Parse the insight_type from BillingAccountLocationInsightTypeInsight resource.
     *
     * @param {string} billingAccountLocationInsightTypeInsightName
     *   A fully-qualified path representing billing_account_location_insight_type_insight resource.
     * @returns {string} A string representing the insight_type.
     */
    matchInsightTypeFromBillingAccountLocationInsightTypeInsightName(billingAccountLocationInsightTypeInsightName) {
        return this.pathTemplates.billingAccountLocationInsightTypeInsightPathTemplate.match(billingAccountLocationInsightTypeInsightName).insight_type;
    }
    /**
     * Parse the insight from BillingAccountLocationInsightTypeInsight resource.
     *
     * @param {string} billingAccountLocationInsightTypeInsightName
     *   A fully-qualified path representing billing_account_location_insight_type_insight resource.
     * @returns {string} A string representing the insight.
     */
    matchInsightFromBillingAccountLocationInsightTypeInsightName(billingAccountLocationInsightTypeInsightName) {
        return this.pathTemplates.billingAccountLocationInsightTypeInsightPathTemplate.match(billingAccountLocationInsightTypeInsightName).insight;
    }
    /**
     * Return a fully-qualified billingAccountLocationRecommender resource name string.
     *
     * @param {string} billing_account
     * @param {string} location
     * @param {string} recommender
     * @returns {string} Resource name string.
     */
    billingAccountLocationRecommenderPath(billingAccount, location, recommender) {
        return this.pathTemplates.billingAccountLocationRecommenderPathTemplate.render({
            billing_account: billingAccount,
            location: location,
            recommender: recommender,
        });
    }
    /**
     * Parse the billing_account from BillingAccountLocationRecommender resource.
     *
     * @param {string} billingAccountLocationRecommenderName
     *   A fully-qualified path representing billing_account_location_recommender resource.
     * @returns {string} A string representing the billing_account.
     */
    matchBillingAccountFromBillingAccountLocationRecommenderName(billingAccountLocationRecommenderName) {
        return this.pathTemplates.billingAccountLocationRecommenderPathTemplate.match(billingAccountLocationRecommenderName).billing_account;
    }
    /**
     * Parse the location from BillingAccountLocationRecommender resource.
     *
     * @param {string} billingAccountLocationRecommenderName
     *   A fully-qualified path representing billing_account_location_recommender resource.
     * @returns {string} A string representing the location.
     */
    matchLocationFromBillingAccountLocationRecommenderName(billingAccountLocationRecommenderName) {
        return this.pathTemplates.billingAccountLocationRecommenderPathTemplate.match(billingAccountLocationRecommenderName).location;
    }
    /**
     * Parse the recommender from BillingAccountLocationRecommender resource.
     *
     * @param {string} billingAccountLocationRecommenderName
     *   A fully-qualified path representing billing_account_location_recommender resource.
     * @returns {string} A string representing the recommender.
     */
    matchRecommenderFromBillingAccountLocationRecommenderName(billingAccountLocationRecommenderName) {
        return this.pathTemplates.billingAccountLocationRecommenderPathTemplate.match(billingAccountLocationRecommenderName).recommender;
    }
    /**
     * Return a fully-qualified billingAccountLocationRecommenderConfig resource name string.
     *
     * @param {string} billing_account
     * @param {string} location
     * @param {string} recommender
     * @returns {string} Resource name string.
     */
    billingAccountLocationRecommenderConfigPath(billingAccount, location, recommender) {
        return this.pathTemplates.billingAccountLocationRecommenderConfigPathTemplate.render({
            billing_account: billingAccount,
            location: location,
            recommender: recommender,
        });
    }
    /**
     * Parse the billing_account from BillingAccountLocationRecommenderConfig resource.
     *
     * @param {string} billingAccountLocationRecommenderConfigName
     *   A fully-qualified path representing billing_account_location_recommender_config resource.
     * @returns {string} A string representing the billing_account.
     */
    matchBillingAccountFromBillingAccountLocationRecommenderConfigName(billingAccountLocationRecommenderConfigName) {
        return this.pathTemplates.billingAccountLocationRecommenderConfigPathTemplate.match(billingAccountLocationRecommenderConfigName).billing_account;
    }
    /**
     * Parse the location from BillingAccountLocationRecommenderConfig resource.
     *
     * @param {string} billingAccountLocationRecommenderConfigName
     *   A fully-qualified path representing billing_account_location_recommender_config resource.
     * @returns {string} A string representing the location.
     */
    matchLocationFromBillingAccountLocationRecommenderConfigName(billingAccountLocationRecommenderConfigName) {
        return this.pathTemplates.billingAccountLocationRecommenderConfigPathTemplate.match(billingAccountLocationRecommenderConfigName).location;
    }
    /**
     * Parse the recommender from BillingAccountLocationRecommenderConfig resource.
     *
     * @param {string} billingAccountLocationRecommenderConfigName
     *   A fully-qualified path representing billing_account_location_recommender_config resource.
     * @returns {string} A string representing the recommender.
     */
    matchRecommenderFromBillingAccountLocationRecommenderConfigName(billingAccountLocationRecommenderConfigName) {
        return this.pathTemplates.billingAccountLocationRecommenderConfigPathTemplate.match(billingAccountLocationRecommenderConfigName).recommender;
    }
    /**
     * Return a fully-qualified billingAccountLocationRecommenderRecommendation resource name string.
     *
     * @param {string} billing_account
     * @param {string} location
     * @param {string} recommender
     * @param {string} recommendation
     * @returns {string} Resource name string.
     */
    billingAccountLocationRecommenderRecommendationPath(billingAccount, location, recommender, recommendation) {
        return this.pathTemplates.billingAccountLocationRecommenderRecommendationPathTemplate.render({
            billing_account: billingAccount,
            location: location,
            recommender: recommender,
            recommendation: recommendation,
        });
    }
    /**
     * Parse the billing_account from BillingAccountLocationRecommenderRecommendation resource.
     *
     * @param {string} billingAccountLocationRecommenderRecommendationName
     *   A fully-qualified path representing billing_account_location_recommender_recommendation resource.
     * @returns {string} A string representing the billing_account.
     */
    matchBillingAccountFromBillingAccountLocationRecommenderRecommendationName(billingAccountLocationRecommenderRecommendationName) {
        return this.pathTemplates.billingAccountLocationRecommenderRecommendationPathTemplate.match(billingAccountLocationRecommenderRecommendationName).billing_account;
    }
    /**
     * Parse the location from BillingAccountLocationRecommenderRecommendation resource.
     *
     * @param {string} billingAccountLocationRecommenderRecommendationName
     *   A fully-qualified path representing billing_account_location_recommender_recommendation resource.
     * @returns {string} A string representing the location.
     */
    matchLocationFromBillingAccountLocationRecommenderRecommendationName(billingAccountLocationRecommenderRecommendationName) {
        return this.pathTemplates.billingAccountLocationRecommenderRecommendationPathTemplate.match(billingAccountLocationRecommenderRecommendationName).location;
    }
    /**
     * Parse the recommender from BillingAccountLocationRecommenderRecommendation resource.
     *
     * @param {string} billingAccountLocationRecommenderRecommendationName
     *   A fully-qualified path representing billing_account_location_recommender_recommendation resource.
     * @returns {string} A string representing the recommender.
     */
    matchRecommenderFromBillingAccountLocationRecommenderRecommendationName(billingAccountLocationRecommenderRecommendationName) {
        return this.pathTemplates.billingAccountLocationRecommenderRecommendationPathTemplate.match(billingAccountLocationRecommenderRecommendationName).recommender;
    }
    /**
     * Parse the recommendation from BillingAccountLocationRecommenderRecommendation resource.
     *
     * @param {string} billingAccountLocationRecommenderRecommendationName
     *   A fully-qualified path representing billing_account_location_recommender_recommendation resource.
     * @returns {string} A string representing the recommendation.
     */
    matchRecommendationFromBillingAccountLocationRecommenderRecommendationName(billingAccountLocationRecommenderRecommendationName) {
        return this.pathTemplates.billingAccountLocationRecommenderRecommendationPathTemplate.match(billingAccountLocationRecommenderRecommendationName).recommendation;
    }
    /**
     * Return a fully-qualified folderLocationInsightType resource name string.
     *
     * @param {string} folder
     * @param {string} location
     * @param {string} insight_type
     * @returns {string} Resource name string.
     */
    folderLocationInsightTypePath(folder, location, insightType) {
        return this.pathTemplates.folderLocationInsightTypePathTemplate.render({
            folder: folder,
            location: location,
            insight_type: insightType,
        });
    }
    /**
     * Parse the folder from FolderLocationInsightType resource.
     *
     * @param {string} folderLocationInsightTypeName
     *   A fully-qualified path representing folder_location_insight_type resource.
     * @returns {string} A string representing the folder.
     */
    matchFolderFromFolderLocationInsightTypeName(folderLocationInsightTypeName) {
        return this.pathTemplates.folderLocationInsightTypePathTemplate.match(folderLocationInsightTypeName).folder;
    }
    /**
     * Parse the location from FolderLocationInsightType resource.
     *
     * @param {string} folderLocationInsightTypeName
     *   A fully-qualified path representing folder_location_insight_type resource.
     * @returns {string} A string representing the location.
     */
    matchLocationFromFolderLocationInsightTypeName(folderLocationInsightTypeName) {
        return this.pathTemplates.folderLocationInsightTypePathTemplate.match(folderLocationInsightTypeName).location;
    }
    /**
     * Parse the insight_type from FolderLocationInsightType resource.
     *
     * @param {string} folderLocationInsightTypeName
     *   A fully-qualified path representing folder_location_insight_type resource.
     * @returns {string} A string representing the insight_type.
     */
    matchInsightTypeFromFolderLocationInsightTypeName(folderLocationInsightTypeName) {
        return this.pathTemplates.folderLocationInsightTypePathTemplate.match(folderLocationInsightTypeName).insight_type;
    }
    /**
     * Return a fully-qualified folderLocationInsightTypeInsight resource name string.
     *
     * @param {string} folder
     * @param {string} location
     * @param {string} insight_type
     * @param {string} insight
     * @returns {string} Resource name string.
     */
    folderLocationInsightTypeInsightPath(folder, location, insightType, insight) {
        return this.pathTemplates.folderLocationInsightTypeInsightPathTemplate.render({
            folder: folder,
            location: location,
            insight_type: insightType,
            insight: insight,
        });
    }
    /**
     * Parse the folder from FolderLocationInsightTypeInsight resource.
     *
     * @param {string} folderLocationInsightTypeInsightName
     *   A fully-qualified path representing folder_location_insight_type_insight resource.
     * @returns {string} A string representing the folder.
     */
    matchFolderFromFolderLocationInsightTypeInsightName(folderLocationInsightTypeInsightName) {
        return this.pathTemplates.folderLocationInsightTypeInsightPathTemplate.match(folderLocationInsightTypeInsightName).folder;
    }
    /**
     * Parse the location from FolderLocationInsightTypeInsight resource.
     *
     * @param {string} folderLocationInsightTypeInsightName
     *   A fully-qualified path representing folder_location_insight_type_insight resource.
     * @returns {string} A string representing the location.
     */
    matchLocationFromFolderLocationInsightTypeInsightName(folderLocationInsightTypeInsightName) {
        return this.pathTemplates.folderLocationInsightTypeInsightPathTemplate.match(folderLocationInsightTypeInsightName).location;
    }
    /**
     * Parse the insight_type from FolderLocationInsightTypeInsight resource.
     *
     * @param {string} folderLocationInsightTypeInsightName
     *   A fully-qualified path representing folder_location_insight_type_insight resource.
     * @returns {string} A string representing the insight_type.
     */
    matchInsightTypeFromFolderLocationInsightTypeInsightName(folderLocationInsightTypeInsightName) {
        return this.pathTemplates.folderLocationInsightTypeInsightPathTemplate.match(folderLocationInsightTypeInsightName).insight_type;
    }
    /**
     * Parse the insight from FolderLocationInsightTypeInsight resource.
     *
     * @param {string} folderLocationInsightTypeInsightName
     *   A fully-qualified path representing folder_location_insight_type_insight resource.
     * @returns {string} A string representing the insight.
     */
    matchInsightFromFolderLocationInsightTypeInsightName(folderLocationInsightTypeInsightName) {
        return this.pathTemplates.folderLocationInsightTypeInsightPathTemplate.match(folderLocationInsightTypeInsightName).insight;
    }
    /**
     * Return a fully-qualified folderLocationRecommender resource name string.
     *
     * @param {string} folder
     * @param {string} location
     * @param {string} recommender
     * @returns {string} Resource name string.
     */
    folderLocationRecommenderPath(folder, location, recommender) {
        return this.pathTemplates.folderLocationRecommenderPathTemplate.render({
            folder: folder,
            location: location,
            recommender: recommender,
        });
    }
    /**
     * Parse the folder from FolderLocationRecommender resource.
     *
     * @param {string} folderLocationRecommenderName
     *   A fully-qualified path representing folder_location_recommender resource.
     * @returns {string} A string representing the folder.
     */
    matchFolderFromFolderLocationRecommenderName(folderLocationRecommenderName) {
        return this.pathTemplates.folderLocationRecommenderPathTemplate.match(folderLocationRecommenderName).folder;
    }
    /**
     * Parse the location from FolderLocationRecommender resource.
     *
     * @param {string} folderLocationRecommenderName
     *   A fully-qualified path representing folder_location_recommender resource.
     * @returns {string} A string representing the location.
     */
    matchLocationFromFolderLocationRecommenderName(folderLocationRecommenderName) {
        return this.pathTemplates.folderLocationRecommenderPathTemplate.match(folderLocationRecommenderName).location;
    }
    /**
     * Parse the recommender from FolderLocationRecommender resource.
     *
     * @param {string} folderLocationRecommenderName
     *   A fully-qualified path representing folder_location_recommender resource.
     * @returns {string} A string representing the recommender.
     */
    matchRecommenderFromFolderLocationRecommenderName(folderLocationRecommenderName) {
        return this.pathTemplates.folderLocationRecommenderPathTemplate.match(folderLocationRecommenderName).recommender;
    }
    /**
     * Return a fully-qualified folderLocationRecommenderRecommendation resource name string.
     *
     * @param {string} folder
     * @param {string} location
     * @param {string} recommender
     * @param {string} recommendation
     * @returns {string} Resource name string.
     */
    folderLocationRecommenderRecommendationPath(folder, location, recommender, recommendation) {
        return this.pathTemplates.folderLocationRecommenderRecommendationPathTemplate.render({
            folder: folder,
            location: location,
            recommender: recommender,
            recommendation: recommendation,
        });
    }
    /**
     * Parse the folder from FolderLocationRecommenderRecommendation resource.
     *
     * @param {string} folderLocationRecommenderRecommendationName
     *   A fully-qualified path representing folder_location_recommender_recommendation resource.
     * @returns {string} A string representing the folder.
     */
    matchFolderFromFolderLocationRecommenderRecommendationName(folderLocationRecommenderRecommendationName) {
        return this.pathTemplates.folderLocationRecommenderRecommendationPathTemplate.match(folderLocationRecommenderRecommendationName).folder;
    }
    /**
     * Parse the location from FolderLocationRecommenderRecommendation resource.
     *
     * @param {string} folderLocationRecommenderRecommendationName
     *   A fully-qualified path representing folder_location_recommender_recommendation resource.
     * @returns {string} A string representing the location.
     */
    matchLocationFromFolderLocationRecommenderRecommendationName(folderLocationRecommenderRecommendationName) {
        return this.pathTemplates.folderLocationRecommenderRecommendationPathTemplate.match(folderLocationRecommenderRecommendationName).location;
    }
    /**
     * Parse the recommender from FolderLocationRecommenderRecommendation resource.
     *
     * @param {string} folderLocationRecommenderRecommendationName
     *   A fully-qualified path representing folder_location_recommender_recommendation resource.
     * @returns {string} A string representing the recommender.
     */
    matchRecommenderFromFolderLocationRecommenderRecommendationName(folderLocationRecommenderRecommendationName) {
        return this.pathTemplates.folderLocationRecommenderRecommendationPathTemplate.match(folderLocationRecommenderRecommendationName).recommender;
    }
    /**
     * Parse the recommendation from FolderLocationRecommenderRecommendation resource.
     *
     * @param {string} folderLocationRecommenderRecommendationName
     *   A fully-qualified path representing folder_location_recommender_recommendation resource.
     * @returns {string} A string representing the recommendation.
     */
    matchRecommendationFromFolderLocationRecommenderRecommendationName(folderLocationRecommenderRecommendationName) {
        return this.pathTemplates.folderLocationRecommenderRecommendationPathTemplate.match(folderLocationRecommenderRecommendationName).recommendation;
    }
    /**
     * Return a fully-qualified organizationLocationInsightType resource name string.
     *
     * @param {string} organization
     * @param {string} location
     * @param {string} insight_type
     * @returns {string} Resource name string.
     */
    organizationLocationInsightTypePath(organization, location, insightType) {
        return this.pathTemplates.organizationLocationInsightTypePathTemplate.render({
            organization: organization,
            location: location,
            insight_type: insightType,
        });
    }
    /**
     * Parse the organization from OrganizationLocationInsightType resource.
     *
     * @param {string} organizationLocationInsightTypeName
     *   A fully-qualified path representing organization_location_insight_type resource.
     * @returns {string} A string representing the organization.
     */
    matchOrganizationFromOrganizationLocationInsightTypeName(organizationLocationInsightTypeName) {
        return this.pathTemplates.organizationLocationInsightTypePathTemplate.match(organizationLocationInsightTypeName).organization;
    }
    /**
     * Parse the location from OrganizationLocationInsightType resource.
     *
     * @param {string} organizationLocationInsightTypeName
     *   A fully-qualified path representing organization_location_insight_type resource.
     * @returns {string} A string representing the location.
     */
    matchLocationFromOrganizationLocationInsightTypeName(organizationLocationInsightTypeName) {
        return this.pathTemplates.organizationLocationInsightTypePathTemplate.match(organizationLocationInsightTypeName).location;
    }
    /**
     * Parse the insight_type from OrganizationLocationInsightType resource.
     *
     * @param {string} organizationLocationInsightTypeName
     *   A fully-qualified path representing organization_location_insight_type resource.
     * @returns {string} A string representing the insight_type.
     */
    matchInsightTypeFromOrganizationLocationInsightTypeName(organizationLocationInsightTypeName) {
        return this.pathTemplates.organizationLocationInsightTypePathTemplate.match(organizationLocationInsightTypeName).insight_type;
    }
    /**
     * Return a fully-qualified organizationLocationInsightTypeConfig resource name string.
     *
     * @param {string} organization
     * @param {string} location
     * @param {string} insight_type
     * @returns {string} Resource name string.
     */
    organizationLocationInsightTypeConfigPath(organization, location, insightType) {
        return this.pathTemplates.organizationLocationInsightTypeConfigPathTemplate.render({
            organization: organization,
            location: location,
            insight_type: insightType,
        });
    }
    /**
     * Parse the organization from OrganizationLocationInsightTypeConfig resource.
     *
     * @param {string} organizationLocationInsightTypeConfigName
     *   A fully-qualified path representing organization_location_insight_type_config resource.
     * @returns {string} A string representing the organization.
     */
    matchOrganizationFromOrganizationLocationInsightTypeConfigName(organizationLocationInsightTypeConfigName) {
        return this.pathTemplates.organizationLocationInsightTypeConfigPathTemplate.match(organizationLocationInsightTypeConfigName).organization;
    }
    /**
     * Parse the location from OrganizationLocationInsightTypeConfig resource.
     *
     * @param {string} organizationLocationInsightTypeConfigName
     *   A fully-qualified path representing organization_location_insight_type_config resource.
     * @returns {string} A string representing the location.
     */
    matchLocationFromOrganizationLocationInsightTypeConfigName(organizationLocationInsightTypeConfigName) {
        return this.pathTemplates.organizationLocationInsightTypeConfigPathTemplate.match(organizationLocationInsightTypeConfigName).location;
    }
    /**
     * Parse the insight_type from OrganizationLocationInsightTypeConfig resource.
     *
     * @param {string} organizationLocationInsightTypeConfigName
     *   A fully-qualified path representing organization_location_insight_type_config resource.
     * @returns {string} A string representing the insight_type.
     */
    matchInsightTypeFromOrganizationLocationInsightTypeConfigName(organizationLocationInsightTypeConfigName) {
        return this.pathTemplates.organizationLocationInsightTypeConfigPathTemplate.match(organizationLocationInsightTypeConfigName).insight_type;
    }
    /**
     * Return a fully-qualified organizationLocationInsightTypeInsight resource name string.
     *
     * @param {string} organization
     * @param {string} location
     * @param {string} insight_type
     * @param {string} insight
     * @returns {string} Resource name string.
     */
    organizationLocationInsightTypeInsightPath(organization, location, insightType, insight) {
        return this.pathTemplates.organizationLocationInsightTypeInsightPathTemplate.render({
            organization: organization,
            location: location,
            insight_type: insightType,
            insight: insight,
        });
    }
    /**
     * Parse the organization from OrganizationLocationInsightTypeInsight resource.
     *
     * @param {string} organizationLocationInsightTypeInsightName
     *   A fully-qualified path representing organization_location_insight_type_insight resource.
     * @returns {string} A string representing the organization.
     */
    matchOrganizationFromOrganizationLocationInsightTypeInsightName(organizationLocationInsightTypeInsightName) {
        return this.pathTemplates.organizationLocationInsightTypeInsightPathTemplate.match(organizationLocationInsightTypeInsightName).organization;
    }
    /**
     * Parse the location from OrganizationLocationInsightTypeInsight resource.
     *
     * @param {string} organizationLocationInsightTypeInsightName
     *   A fully-qualified path representing organization_location_insight_type_insight resource.
     * @returns {string} A string representing the location.
     */
    matchLocationFromOrganizationLocationInsightTypeInsightName(organizationLocationInsightTypeInsightName) {
        return this.pathTemplates.organizationLocationInsightTypeInsightPathTemplate.match(organizationLocationInsightTypeInsightName).location;
    }
    /**
     * Parse the insight_type from OrganizationLocationInsightTypeInsight resource.
     *
     * @param {string} organizationLocationInsightTypeInsightName
     *   A fully-qualified path representing organization_location_insight_type_insight resource.
     * @returns {string} A string representing the insight_type.
     */
    matchInsightTypeFromOrganizationLocationInsightTypeInsightName(organizationLocationInsightTypeInsightName) {
        return this.pathTemplates.organizationLocationInsightTypeInsightPathTemplate.match(organizationLocationInsightTypeInsightName).insight_type;
    }
    /**
     * Parse the insight from OrganizationLocationInsightTypeInsight resource.
     *
     * @param {string} organizationLocationInsightTypeInsightName
     *   A fully-qualified path representing organization_location_insight_type_insight resource.
     * @returns {string} A string representing the insight.
     */
    matchInsightFromOrganizationLocationInsightTypeInsightName(organizationLocationInsightTypeInsightName) {
        return this.pathTemplates.organizationLocationInsightTypeInsightPathTemplate.match(organizationLocationInsightTypeInsightName).insight;
    }
    /**
     * Return a fully-qualified organizationLocationRecommender resource name string.
     *
     * @param {string} organization
     * @param {string} location
     * @param {string} recommender
     * @returns {string} Resource name string.
     */
    organizationLocationRecommenderPath(organization, location, recommender) {
        return this.pathTemplates.organizationLocationRecommenderPathTemplate.render({
            organization: organization,
            location: location,
            recommender: recommender,
        });
    }
    /**
     * Parse the organization from OrganizationLocationRecommender resource.
     *
     * @param {string} organizationLocationRecommenderName
     *   A fully-qualified path representing organization_location_recommender resource.
     * @returns {string} A string representing the organization.
     */
    matchOrganizationFromOrganizationLocationRecommenderName(organizationLocationRecommenderName) {
        return this.pathTemplates.organizationLocationRecommenderPathTemplate.match(organizationLocationRecommenderName).organization;
    }
    /**
     * Parse the location from OrganizationLocationRecommender resource.
     *
     * @param {string} organizationLocationRecommenderName
     *   A fully-qualified path representing organization_location_recommender resource.
     * @returns {string} A string representing the location.
     */
    matchLocationFromOrganizationLocationRecommenderName(organizationLocationRecommenderName) {
        return this.pathTemplates.organizationLocationRecommenderPathTemplate.match(organizationLocationRecommenderName).location;
    }
    /**
     * Parse the recommender from OrganizationLocationRecommender resource.
     *
     * @param {string} organizationLocationRecommenderName
     *   A fully-qualified path representing organization_location_recommender resource.
     * @returns {string} A string representing the recommender.
     */
    matchRecommenderFromOrganizationLocationRecommenderName(organizationLocationRecommenderName) {
        return this.pathTemplates.organizationLocationRecommenderPathTemplate.match(organizationLocationRecommenderName).recommender;
    }
    /**
     * Return a fully-qualified organizationLocationRecommenderConfig resource name string.
     *
     * @param {string} organization
     * @param {string} location
     * @param {string} recommender
     * @returns {string} Resource name string.
     */
    organizationLocationRecommenderConfigPath(organization, location, recommender) {
        return this.pathTemplates.organizationLocationRecommenderConfigPathTemplate.render({
            organization: organization,
            location: location,
            recommender: recommender,
        });
    }
    /**
     * Parse the organization from OrganizationLocationRecommenderConfig resource.
     *
     * @param {string} organizationLocationRecommenderConfigName
     *   A fully-qualified path representing organization_location_recommender_config resource.
     * @returns {string} A string representing the organization.
     */
    matchOrganizationFromOrganizationLocationRecommenderConfigName(organizationLocationRecommenderConfigName) {
        return this.pathTemplates.organizationLocationRecommenderConfigPathTemplate.match(organizationLocationRecommenderConfigName).organization;
    }
    /**
     * Parse the location from OrganizationLocationRecommenderConfig resource.
     *
     * @param {string} organizationLocationRecommenderConfigName
     *   A fully-qualified path representing organization_location_recommender_config resource.
     * @returns {string} A string representing the location.
     */
    matchLocationFromOrganizationLocationRecommenderConfigName(organizationLocationRecommenderConfigName) {
        return this.pathTemplates.organizationLocationRecommenderConfigPathTemplate.match(organizationLocationRecommenderConfigName).location;
    }
    /**
     * Parse the recommender from OrganizationLocationRecommenderConfig resource.
     *
     * @param {string} organizationLocationRecommenderConfigName
     *   A fully-qualified path representing organization_location_recommender_config resource.
     * @returns {string} A string representing the recommender.
     */
    matchRecommenderFromOrganizationLocationRecommenderConfigName(organizationLocationRecommenderConfigName) {
        return this.pathTemplates.organizationLocationRecommenderConfigPathTemplate.match(organizationLocationRecommenderConfigName).recommender;
    }
    /**
     * Return a fully-qualified organizationLocationRecommenderRecommendation resource name string.
     *
     * @param {string} organization
     * @param {string} location
     * @param {string} recommender
     * @param {string} recommendation
     * @returns {string} Resource name string.
     */
    organizationLocationRecommenderRecommendationPath(organization, location, recommender, recommendation) {
        return this.pathTemplates.organizationLocationRecommenderRecommendationPathTemplate.render({
            organization: organization,
            location: location,
            recommender: recommender,
            recommendation: recommendation,
        });
    }
    /**
     * Parse the organization from OrganizationLocationRecommenderRecommendation resource.
     *
     * @param {string} organizationLocationRecommenderRecommendationName
     *   A fully-qualified path representing organization_location_recommender_recommendation resource.
     * @returns {string} A string representing the organization.
     */
    matchOrganizationFromOrganizationLocationRecommenderRecommendationName(organizationLocationRecommenderRecommendationName) {
        return this.pathTemplates.organizationLocationRecommenderRecommendationPathTemplate.match(organizationLocationRecommenderRecommendationName).organization;
    }
    /**
     * Parse the location from OrganizationLocationRecommenderRecommendation resource.
     *
     * @param {string} organizationLocationRecommenderRecommendationName
     *   A fully-qualified path representing organization_location_recommender_recommendation resource.
     * @returns {string} A string representing the location.
     */
    matchLocationFromOrganizationLocationRecommenderRecommendationName(organizationLocationRecommenderRecommendationName) {
        return this.pathTemplates.organizationLocationRecommenderRecommendationPathTemplate.match(organizationLocationRecommenderRecommendationName).location;
    }
    /**
     * Parse the recommender from OrganizationLocationRecommenderRecommendation resource.
     *
     * @param {string} organizationLocationRecommenderRecommendationName
     *   A fully-qualified path representing organization_location_recommender_recommendation resource.
     * @returns {string} A string representing the recommender.
     */
    matchRecommenderFromOrganizationLocationRecommenderRecommendationName(organizationLocationRecommenderRecommendationName) {
        return this.pathTemplates.organizationLocationRecommenderRecommendationPathTemplate.match(organizationLocationRecommenderRecommendationName).recommender;
    }
    /**
     * Parse the recommendation from OrganizationLocationRecommenderRecommendation resource.
     *
     * @param {string} organizationLocationRecommenderRecommendationName
     *   A fully-qualified path representing organization_location_recommender_recommendation resource.
     * @returns {string} A string representing the recommendation.
     */
    matchRecommendationFromOrganizationLocationRecommenderRecommendationName(organizationLocationRecommenderRecommendationName) {
        return this.pathTemplates.organizationLocationRecommenderRecommendationPathTemplate.match(organizationLocationRecommenderRecommendationName).recommendation;
    }
    /**
     * Return a fully-qualified projectLocationInsightType resource name string.
     *
     * @param {string} project
     * @param {string} location
     * @param {string} insight_type
     * @returns {string} Resource name string.
     */
    projectLocationInsightTypePath(project, location, insightType) {
        return this.pathTemplates.projectLocationInsightTypePathTemplate.render({
            project: project,
            location: location,
            insight_type: insightType,
        });
    }
    /**
     * Parse the project from ProjectLocationInsightType resource.
     *
     * @param {string} projectLocationInsightTypeName
     *   A fully-qualified path representing project_location_insight_type resource.
     * @returns {string} A string representing the project.
     */
    matchProjectFromProjectLocationInsightTypeName(projectLocationInsightTypeName) {
        return this.pathTemplates.projectLocationInsightTypePathTemplate.match(projectLocationInsightTypeName).project;
    }
    /**
     * Parse the location from ProjectLocationInsightType resource.
     *
     * @param {string} projectLocationInsightTypeName
     *   A fully-qualified path representing project_location_insight_type resource.
     * @returns {string} A string representing the location.
     */
    matchLocationFromProjectLocationInsightTypeName(projectLocationInsightTypeName) {
        return this.pathTemplates.projectLocationInsightTypePathTemplate.match(projectLocationInsightTypeName).location;
    }
    /**
     * Parse the insight_type from ProjectLocationInsightType resource.
     *
     * @param {string} projectLocationInsightTypeName
     *   A fully-qualified path representing project_location_insight_type resource.
     * @returns {string} A string representing the insight_type.
     */
    matchInsightTypeFromProjectLocationInsightTypeName(projectLocationInsightTypeName) {
        return this.pathTemplates.projectLocationInsightTypePathTemplate.match(projectLocationInsightTypeName).insight_type;
    }
    /**
     * Return a fully-qualified projectLocationInsightTypeConfig resource name string.
     *
     * @param {string} project
     * @param {string} location
     * @param {string} insight_type
     * @returns {string} Resource name string.
     */
    projectLocationInsightTypeConfigPath(project, location, insightType) {
        return this.pathTemplates.projectLocationInsightTypeConfigPathTemplate.render({
            project: project,
            location: location,
            insight_type: insightType,
        });
    }
    /**
     * Parse the project from ProjectLocationInsightTypeConfig resource.
     *
     * @param {string} projectLocationInsightTypeConfigName
     *   A fully-qualified path representing project_location_insight_type_config resource.
     * @returns {string} A string representing the project.
     */
    matchProjectFromProjectLocationInsightTypeConfigName(projectLocationInsightTypeConfigName) {
        return this.pathTemplates.projectLocationInsightTypeConfigPathTemplate.match(projectLocationInsightTypeConfigName).project;
    }
    /**
     * Parse the location from ProjectLocationInsightTypeConfig resource.
     *
     * @param {string} projectLocationInsightTypeConfigName
     *   A fully-qualified path representing project_location_insight_type_config resource.
     * @returns {string} A string representing the location.
     */
    matchLocationFromProjectLocationInsightTypeConfigName(projectLocationInsightTypeConfigName) {
        return this.pathTemplates.projectLocationInsightTypeConfigPathTemplate.match(projectLocationInsightTypeConfigName).location;
    }
    /**
     * Parse the insight_type from ProjectLocationInsightTypeConfig resource.
     *
     * @param {string} projectLocationInsightTypeConfigName
     *   A fully-qualified path representing project_location_insight_type_config resource.
     * @returns {string} A string representing the insight_type.
     */
    matchInsightTypeFromProjectLocationInsightTypeConfigName(projectLocationInsightTypeConfigName) {
        return this.pathTemplates.projectLocationInsightTypeConfigPathTemplate.match(projectLocationInsightTypeConfigName).insight_type;
    }
    /**
     * Return a fully-qualified projectLocationInsightTypeInsight resource name string.
     *
     * @param {string} project
     * @param {string} location
     * @param {string} insight_type
     * @param {string} insight
     * @returns {string} Resource name string.
     */
    projectLocationInsightTypeInsightPath(project, location, insightType, insight) {
        return this.pathTemplates.projectLocationInsightTypeInsightPathTemplate.render({
            project: project,
            location: location,
            insight_type: insightType,
            insight: insight,
        });
    }
    /**
     * Parse the project from ProjectLocationInsightTypeInsight resource.
     *
     * @param {string} projectLocationInsightTypeInsightName
     *   A fully-qualified path representing project_location_insight_type_insight resource.
     * @returns {string} A string representing the project.
     */
    matchProjectFromProjectLocationInsightTypeInsightName(projectLocationInsightTypeInsightName) {
        return this.pathTemplates.projectLocationInsightTypeInsightPathTemplate.match(projectLocationInsightTypeInsightName).project;
    }
    /**
     * Parse the location from ProjectLocationInsightTypeInsight resource.
     *
     * @param {string} projectLocationInsightTypeInsightName
     *   A fully-qualified path representing project_location_insight_type_insight resource.
     * @returns {string} A string representing the location.
     */
    matchLocationFromProjectLocationInsightTypeInsightName(projectLocationInsightTypeInsightName) {
        return this.pathTemplates.projectLocationInsightTypeInsightPathTemplate.match(projectLocationInsightTypeInsightName).location;
    }
    /**
     * Parse the insight_type from ProjectLocationInsightTypeInsight resource.
     *
     * @param {string} projectLocationInsightTypeInsightName
     *   A fully-qualified path representing project_location_insight_type_insight resource.
     * @returns {string} A string representing the insight_type.
     */
    matchInsightTypeFromProjectLocationInsightTypeInsightName(projectLocationInsightTypeInsightName) {
        return this.pathTemplates.projectLocationInsightTypeInsightPathTemplate.match(projectLocationInsightTypeInsightName).insight_type;
    }
    /**
     * Parse the insight from ProjectLocationInsightTypeInsight resource.
     *
     * @param {string} projectLocationInsightTypeInsightName
     *   A fully-qualified path representing project_location_insight_type_insight resource.
     * @returns {string} A string representing the insight.
     */
    matchInsightFromProjectLocationInsightTypeInsightName(projectLocationInsightTypeInsightName) {
        return this.pathTemplates.projectLocationInsightTypeInsightPathTemplate.match(projectLocationInsightTypeInsightName).insight;
    }
    /**
     * Return a fully-qualified projectLocationRecommender resource name string.
     *
     * @param {string} project
     * @param {string} location
     * @param {string} recommender
     * @returns {string} Resource name string.
     */
    projectLocationRecommenderPath(project, location, recommender) {
        return this.pathTemplates.projectLocationRecommenderPathTemplate.render({
            project: project,
            location: location,
            recommender: recommender,
        });
    }
    /**
     * Parse the project from ProjectLocationRecommender resource.
     *
     * @param {string} projectLocationRecommenderName
     *   A fully-qualified path representing project_location_recommender resource.
     * @returns {string} A string representing the project.
     */
    matchProjectFromProjectLocationRecommenderName(projectLocationRecommenderName) {
        return this.pathTemplates.projectLocationRecommenderPathTemplate.match(projectLocationRecommenderName).project;
    }
    /**
     * Parse the location from ProjectLocationRecommender resource.
     *
     * @param {string} projectLocationRecommenderName
     *   A fully-qualified path representing project_location_recommender resource.
     * @returns {string} A string representing the location.
     */
    matchLocationFromProjectLocationRecommenderName(projectLocationRecommenderName) {
        return this.pathTemplates.projectLocationRecommenderPathTemplate.match(projectLocationRecommenderName).location;
    }
    /**
     * Parse the recommender from ProjectLocationRecommender resource.
     *
     * @param {string} projectLocationRecommenderName
     *   A fully-qualified path representing project_location_recommender resource.
     * @returns {string} A string representing the recommender.
     */
    matchRecommenderFromProjectLocationRecommenderName(projectLocationRecommenderName) {
        return this.pathTemplates.projectLocationRecommenderPathTemplate.match(projectLocationRecommenderName).recommender;
    }
    /**
     * Return a fully-qualified projectLocationRecommenderConfig resource name string.
     *
     * @param {string} project
     * @param {string} location
     * @param {string} recommender
     * @returns {string} Resource name string.
     */
    projectLocationRecommenderConfigPath(project, location, recommender) {
        return this.pathTemplates.projectLocationRecommenderConfigPathTemplate.render({
            project: project,
            location: location,
            recommender: recommender,
        });
    }
    /**
     * Parse the project from ProjectLocationRecommenderConfig resource.
     *
     * @param {string} projectLocationRecommenderConfigName
     *   A fully-qualified path representing project_location_recommender_config resource.
     * @returns {string} A string representing the project.
     */
    matchProjectFromProjectLocationRecommenderConfigName(projectLocationRecommenderConfigName) {
        return this.pathTemplates.projectLocationRecommenderConfigPathTemplate.match(projectLocationRecommenderConfigName).project;
    }
    /**
     * Parse the location from ProjectLocationRecommenderConfig resource.
     *
     * @param {string} projectLocationRecommenderConfigName
     *   A fully-qualified path representing project_location_recommender_config resource.
     * @returns {string} A string representing the location.
     */
    matchLocationFromProjectLocationRecommenderConfigName(projectLocationRecommenderConfigName) {
        return this.pathTemplates.projectLocationRecommenderConfigPathTemplate.match(projectLocationRecommenderConfigName).location;
    }
    /**
     * Parse the recommender from ProjectLocationRecommenderConfig resource.
     *
     * @param {string} projectLocationRecommenderConfigName
     *   A fully-qualified path representing project_location_recommender_config resource.
     * @returns {string} A string representing the recommender.
     */
    matchRecommenderFromProjectLocationRecommenderConfigName(projectLocationRecommenderConfigName) {
        return this.pathTemplates.projectLocationRecommenderConfigPathTemplate.match(projectLocationRecommenderConfigName).recommender;
    }
    /**
     * Return a fully-qualified projectLocationRecommenderRecommendation resource name string.
     *
     * @param {string} project
     * @param {string} location
     * @param {string} recommender
     * @param {string} recommendation
     * @returns {string} Resource name string.
     */
    projectLocationRecommenderRecommendationPath(project, location, recommender, recommendation) {
        return this.pathTemplates.projectLocationRecommenderRecommendationPathTemplate.render({
            project: project,
            location: location,
            recommender: recommender,
            recommendation: recommendation,
        });
    }
    /**
     * Parse the project from ProjectLocationRecommenderRecommendation resource.
     *
     * @param {string} projectLocationRecommenderRecommendationName
     *   A fully-qualified path representing project_location_recommender_recommendation resource.
     * @returns {string} A string representing the project.
     */
    matchProjectFromProjectLocationRecommenderRecommendationName(projectLocationRecommenderRecommendationName) {
        return this.pathTemplates.projectLocationRecommenderRecommendationPathTemplate.match(projectLocationRecommenderRecommendationName).project;
    }
    /**
     * Parse the location from ProjectLocationRecommenderRecommendation resource.
     *
     * @param {string} projectLocationRecommenderRecommendationName
     *   A fully-qualified path representing project_location_recommender_recommendation resource.
     * @returns {string} A string representing the location.
     */
    matchLocationFromProjectLocationRecommenderRecommendationName(projectLocationRecommenderRecommendationName) {
        return this.pathTemplates.projectLocationRecommenderRecommendationPathTemplate.match(projectLocationRecommenderRecommendationName).location;
    }
    /**
     * Parse the recommender from ProjectLocationRecommenderRecommendation resource.
     *
     * @param {string} projectLocationRecommenderRecommendationName
     *   A fully-qualified path representing project_location_recommender_recommendation resource.
     * @returns {string} A string representing the recommender.
     */
    matchRecommenderFromProjectLocationRecommenderRecommendationName(projectLocationRecommenderRecommendationName) {
        return this.pathTemplates.projectLocationRecommenderRecommendationPathTemplate.match(projectLocationRecommenderRecommendationName).recommender;
    }
    /**
     * Parse the recommendation from ProjectLocationRecommenderRecommendation resource.
     *
     * @param {string} projectLocationRecommenderRecommendationName
     *   A fully-qualified path representing project_location_recommender_recommendation resource.
     * @returns {string} A string representing the recommendation.
     */
    matchRecommendationFromProjectLocationRecommenderRecommendationName(projectLocationRecommenderRecommendationName) {
        return this.pathTemplates.projectLocationRecommenderRecommendationPathTemplate.match(projectLocationRecommenderRecommendationName).recommendation;
    }
    /**
     * Terminate the gRPC channel and close the client.
     *
     * The client will no longer be usable and all future behavior is undefined.
     * @returns {Promise} A promise that resolves when the client is closed.
     */
    close() {
        if (this.recommenderStub && !this._terminated) {
            return this.recommenderStub.then(stub => {
                this._log.info('ending gRPC channel');
                this._terminated = true;
                stub.close();
            });
        }
        return Promise.resolve();
    }
}
exports.RecommenderClient = RecommenderClient;
