// @aws-sdk/client-compute-optimizer/dist-cjs/index.js
'use strict';

var middlewareHostHeader = require('@aws-sdk/middleware-host-header');
var middlewareLogger = require('@aws-sdk/middleware-logger');
var middlewareRecursionDetection = require('@aws-sdk/middleware-recursion-detection');
var middlewareUserAgent = require('@aws-sdk/middleware-user-agent');
var configResolver = require('@smithy/config-resolver');
var core = require('@smithy/core');
var schema = require('@smithy/core/schema');
var middlewareContentLength = require('@smithy/middleware-content-length');
var middlewareEndpoint = require('@smithy/middleware-endpoint');
var middlewareRetry = require('@smithy/middleware-retry');
var smithyClient = require('@smithy/smithy-client');
var httpAuthSchemeProvider = require('./auth/httpAuthSchemeProvider');
var runtimeConfig = require('./runtimeConfig');
var regionConfigResolver = require('@aws-sdk/region-config-resolver');
var protocolHttp = require('@smithy/protocol-http');
var schemas_0 = require('./schemas/schemas_0');
var errors = require('./models/errors');
var ComputeOptimizerServiceException = require('./models/ComputeOptimizerServiceException');

const resolveClientEndpointParameters = (options) => {
    return Object.assign(options, {
        useDualstackEndpoint: options.useDualstackEndpoint ?? false,
        useFipsEndpoint: options.useFipsEndpoint ?? false,
        defaultSigningName: "compute-optimizer",
    });
};
const commonParams = {
    UseFIPS: { type: "builtInParams", name: "useFipsEndpoint" },
    Endpoint: { type: "builtInParams", name: "endpoint" },
    Region: { type: "builtInParams", name: "region" },
    UseDualStack: { type: "builtInParams", name: "useDualstackEndpoint" },
};

const getHttpAuthExtensionConfiguration = (runtimeConfig) => {
    const _httpAuthSchemes = runtimeConfig.httpAuthSchemes;
    let _httpAuthSchemeProvider = runtimeConfig.httpAuthSchemeProvider;
    let _credentials = runtimeConfig.credentials;
    return {
        setHttpAuthScheme(httpAuthScheme) {
            const index = _httpAuthSchemes.findIndex((scheme) => scheme.schemeId === httpAuthScheme.schemeId);
            if (index === -1) {
                _httpAuthSchemes.push(httpAuthScheme);
            }
            else {
                _httpAuthSchemes.splice(index, 1, httpAuthScheme);
            }
        },
        httpAuthSchemes() {
            return _httpAuthSchemes;
        },
        setHttpAuthSchemeProvider(httpAuthSchemeProvider) {
            _httpAuthSchemeProvider = httpAuthSchemeProvider;
        },
        httpAuthSchemeProvider() {
            return _httpAuthSchemeProvider;
        },
        setCredentials(credentials) {
            _credentials = credentials;
        },
        credentials() {
            return _credentials;
        },
    };
};
const resolveHttpAuthRuntimeConfig = (config) => {
    return {
        httpAuthSchemes: config.httpAuthSchemes(),
        httpAuthSchemeProvider: config.httpAuthSchemeProvider(),
        credentials: config.credentials(),
    };
};

const resolveRuntimeExtensions = (runtimeConfig, extensions) => {
    const extensionConfiguration = Object.assign(regionConfigResolver.getAwsRegionExtensionConfiguration(runtimeConfig), smithyClient.getDefaultExtensionConfiguration(runtimeConfig), protocolHttp.getHttpHandlerExtensionConfiguration(runtimeConfig), getHttpAuthExtensionConfiguration(runtimeConfig));
    extensions.forEach((extension) => extension.configure(extensionConfiguration));
    return Object.assign(runtimeConfig, regionConfigResolver.resolveAwsRegionExtensionConfiguration(extensionConfiguration), smithyClient.resolveDefaultRuntimeConfig(extensionConfiguration), protocolHttp.resolveHttpHandlerRuntimeConfig(extensionConfiguration), resolveHttpAuthRuntimeConfig(extensionConfiguration));
};

class ComputeOptimizerClient extends smithyClient.Client {
    config;
    constructor(...[configuration]) {
        const _config_0 = runtimeConfig.getRuntimeConfig(configuration || {});
        super(_config_0);
        this.initConfig = _config_0;
        const _config_1 = resolveClientEndpointParameters(_config_0);
        const _config_2 = middlewareUserAgent.resolveUserAgentConfig(_config_1);
        const _config_3 = middlewareRetry.resolveRetryConfig(_config_2);
        const _config_4 = configResolver.resolveRegionConfig(_config_3);
        const _config_5 = middlewareHostHeader.resolveHostHeaderConfig(_config_4);
        const _config_6 = middlewareEndpoint.resolveEndpointConfig(_config_5);
        const _config_7 = httpAuthSchemeProvider.resolveHttpAuthSchemeConfig(_config_6);
        const _config_8 = resolveRuntimeExtensions(_config_7, configuration?.extensions || []);
        this.config = _config_8;
        this.middlewareStack.use(schema.getSchemaSerdePlugin(this.config));
        this.middlewareStack.use(middlewareUserAgent.getUserAgentPlugin(this.config));
        this.middlewareStack.use(middlewareRetry.getRetryPlugin(this.config));
        this.middlewareStack.use(middlewareContentLength.getContentLengthPlugin(this.config));
        this.middlewareStack.use(middlewareHostHeader.getHostHeaderPlugin(this.config));
        this.middlewareStack.use(middlewareLogger.getLoggerPlugin(this.config));
        this.middlewareStack.use(middlewareRecursionDetection.getRecursionDetectionPlugin(this.config));
        this.middlewareStack.use(core.getHttpAuthSchemeEndpointRuleSetPlugin(this.config, {
            httpAuthSchemeParametersProvider: httpAuthSchemeProvider.defaultComputeOptimizerHttpAuthSchemeParametersProvider,
            identityProviderConfigProvider: async (config) => new core.DefaultIdentityProviderConfig({
                "aws.auth#sigv4": config.credentials,
            }),
        }));
        this.middlewareStack.use(core.getHttpSigningPlugin(this.config));
    }
    destroy() {
        super.destroy();
    }
}

class DeleteRecommendationPreferencesCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("ComputeOptimizerService", "DeleteRecommendationPreferences", {})
    .n("ComputeOptimizerClient", "DeleteRecommendationPreferencesCommand")
    .sc(schemas_0.DeleteRecommendationPreferences$)
    .build() {
}

class DescribeRecommendationExportJobsCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("ComputeOptimizerService", "DescribeRecommendationExportJobs", {})
    .n("ComputeOptimizerClient", "DescribeRecommendationExportJobsCommand")
    .sc(schemas_0.DescribeRecommendationExportJobs$)
    .build() {
}

class ExportAutoScalingGroupRecommendationsCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("ComputeOptimizerService", "ExportAutoScalingGroupRecommendations", {})
    .n("ComputeOptimizerClient", "ExportAutoScalingGroupRecommendationsCommand")
    .sc(schemas_0.ExportAutoScalingGroupRecommendations$)
    .build() {
}

class ExportEBSVolumeRecommendationsCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("ComputeOptimizerService", "ExportEBSVolumeRecommendations", {})
    .n("ComputeOptimizerClient", "ExportEBSVolumeRecommendationsCommand")
    .sc(schemas_0.ExportEBSVolumeRecommendations$)
    .build() {
}

class ExportEC2InstanceRecommendationsCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("ComputeOptimizerService", "ExportEC2InstanceRecommendations", {})
    .n("ComputeOptimizerClient", "ExportEC2InstanceRecommendationsCommand")
    .sc(schemas_0.ExportEC2InstanceRecommendations$)
    .build() {
}

class ExportECSServiceRecommendationsCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("ComputeOptimizerService", "ExportECSServiceRecommendations", {})
    .n("ComputeOptimizerClient", "ExportECSServiceRecommendationsCommand")
    .sc(schemas_0.ExportECSServiceRecommendations$)
    .build() {
}

class ExportIdleRecommendationsCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("ComputeOptimizerService", "ExportIdleRecommendations", {})
    .n("ComputeOptimizerClient", "ExportIdleRecommendationsCommand")
    .sc(schemas_0.ExportIdleRecommendations$)
    .build() {
}

class ExportLambdaFunctionRecommendationsCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("ComputeOptimizerService", "ExportLambdaFunctionRecommendations", {})
    .n("ComputeOptimizerClient", "ExportLambdaFunctionRecommendationsCommand")
    .sc(schemas_0.ExportLambdaFunctionRecommendations$)
    .build() {
}

class ExportLicenseRecommendationsCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("ComputeOptimizerService", "ExportLicenseRecommendations", {})
    .n("ComputeOptimizerClient", "ExportLicenseRecommendationsCommand")
    .sc(schemas_0.ExportLicenseRecommendations$)
    .build() {
}

class ExportRDSDatabaseRecommendationsCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("ComputeOptimizerService", "ExportRDSDatabaseRecommendations", {})
    .n("ComputeOptimizerClient", "ExportRDSDatabaseRecommendationsCommand")
    .sc(schemas_0.ExportRDSDatabaseRecommendations$)
    .build() {
}

class GetAutoScalingGroupRecommendationsCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("ComputeOptimizerService", "GetAutoScalingGroupRecommendations", {})
    .n("ComputeOptimizerClient", "GetAutoScalingGroupRecommendationsCommand")
    .sc(schemas_0.GetAutoScalingGroupRecommendations$)
    .build() {
}

class GetEBSVolumeRecommendationsCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("ComputeOptimizerService", "GetEBSVolumeRecommendations", {})
    .n("ComputeOptimizerClient", "GetEBSVolumeRecommendationsCommand")
    .sc(schemas_0.GetEBSVolumeRecommendations$)
    .build() {
}

class GetEC2InstanceRecommendationsCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("ComputeOptimizerService", "GetEC2InstanceRecommendations", {})
    .n("ComputeOptimizerClient", "GetEC2InstanceRecommendationsCommand")
    .sc(schemas_0.GetEC2InstanceRecommendations$)
    .build() {
}

class GetEC2RecommendationProjectedMetricsCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("ComputeOptimizerService", "GetEC2RecommendationProjectedMetrics", {})
    .n("ComputeOptimizerClient", "GetEC2RecommendationProjectedMetricsCommand")
    .sc(schemas_0.GetEC2RecommendationProjectedMetrics$)
    .build() {
}

class GetECSServiceRecommendationProjectedMetricsCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("ComputeOptimizerService", "GetECSServiceRecommendationProjectedMetrics", {})
    .n("ComputeOptimizerClient", "GetECSServiceRecommendationProjectedMetricsCommand")
    .sc(schemas_0.GetECSServiceRecommendationProjectedMetrics$)
    .build() {
}

class GetECSServiceRecommendationsCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("ComputeOptimizerService", "GetECSServiceRecommendations", {})
    .n("ComputeOptimizerClient", "GetECSServiceRecommendationsCommand")
    .sc(schemas_0.GetECSServiceRecommendations$)
    .build() {
}

class GetEffectiveRecommendationPreferencesCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("ComputeOptimizerService", "GetEffectiveRecommendationPreferences", {})
    .n("ComputeOptimizerClient", "GetEffectiveRecommendationPreferencesCommand")
    .sc(schemas_0.GetEffectiveRecommendationPreferences$)
    .build() {
}

class GetEnrollmentStatusCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("ComputeOptimizerService", "GetEnrollmentStatus", {})
    .n("ComputeOptimizerClient", "GetEnrollmentStatusCommand")
    .sc(schemas_0.GetEnrollmentStatus$)
    .build() {
}

class GetEnrollmentStatusesForOrganizationCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("ComputeOptimizerService", "GetEnrollmentStatusesForOrganization", {})
    .n("ComputeOptimizerClient", "GetEnrollmentStatusesForOrganizationCommand")
    .sc(schemas_0.GetEnrollmentStatusesForOrganization$)
    .build() {
}

class GetIdleRecommendationsCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("ComputeOptimizerService", "GetIdleRecommendations", {})
    .n("ComputeOptimizerClient", "GetIdleRecommendationsCommand")
    .sc(schemas_0.GetIdleRecommendations$)
    .build() {
}

class GetLambdaFunctionRecommendationsCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("ComputeOptimizerService", "GetLambdaFunctionRecommendations", {})
    .n("ComputeOptimizerClient", "GetLambdaFunctionRecommendationsCommand")
    .sc(schemas_0.GetLambdaFunctionRecommendations$)
    .build() {
}

class GetLicenseRecommendationsCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("ComputeOptimizerService", "GetLicenseRecommendations", {})
    .n("ComputeOptimizerClient", "GetLicenseRecommendationsCommand")
    .sc(schemas_0.GetLicenseRecommendations$)
    .build() {
}

class GetRDSDatabaseRecommendationProjectedMetricsCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("ComputeOptimizerService", "GetRDSDatabaseRecommendationProjectedMetrics", {})
    .n("ComputeOptimizerClient", "GetRDSDatabaseRecommendationProjectedMetricsCommand")
    .sc(schemas_0.GetRDSDatabaseRecommendationProjectedMetrics$)
    .build() {
}

class GetRDSDatabaseRecommendationsCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("ComputeOptimizerService", "GetRDSDatabaseRecommendations", {})
    .n("ComputeOptimizerClient", "GetRDSDatabaseRecommendationsCommand")
    .sc(schemas_0.GetRDSDatabaseRecommendations$)
    .build() {
}

class GetRecommendationPreferencesCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("ComputeOptimizerService", "GetRecommendationPreferences", {})
    .n("ComputeOptimizerClient", "GetRecommendationPreferencesCommand")
    .sc(schemas_0.GetRecommendationPreferences$)
    .build() {
}

class GetRecommendationSummariesCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("ComputeOptimizerService", "GetRecommendationSummaries", {})
    .n("ComputeOptimizerClient", "GetRecommendationSummariesCommand")
    .sc(schemas_0.GetRecommendationSummaries$)
    .build() {
}

class PutRecommendationPreferencesCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("ComputeOptimizerService", "PutRecommendationPreferences", {})
    .n("ComputeOptimizerClient", "PutRecommendationPreferencesCommand")
    .sc(schemas_0.PutRecommendationPreferences$)
    .build() {
}

class UpdateEnrollmentStatusCommand extends smithyClient.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [middlewareEndpoint.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("ComputeOptimizerService", "UpdateEnrollmentStatus", {})
    .n("ComputeOptimizerClient", "UpdateEnrollmentStatusCommand")
    .sc(schemas_0.UpdateEnrollmentStatus$)
    .build() {
}

const paginateDescribeRecommendationExportJobs = core.createPaginator(ComputeOptimizerClient, DescribeRecommendationExportJobsCommand, "nextToken", "nextToken", "maxResults");

const paginateGetEnrollmentStatusesForOrganization = core.createPaginator(ComputeOptimizerClient, GetEnrollmentStatusesForOrganizationCommand, "nextToken", "nextToken", "maxResults");

const paginateGetLambdaFunctionRecommendations = core.createPaginator(ComputeOptimizerClient, GetLambdaFunctionRecommendationsCommand, "nextToken", "nextToken", "maxResults");

const paginateGetRecommendationPreferences = core.createPaginator(ComputeOptimizerClient, GetRecommendationPreferencesCommand, "nextToken", "nextToken", "maxResults");

const paginateGetRecommendationSummaries = core.createPaginator(ComputeOptimizerClient, GetRecommendationSummariesCommand, "nextToken", "nextToken", "maxResults");

const commands = {
    DeleteRecommendationPreferencesCommand,
    DescribeRecommendationExportJobsCommand,
    ExportAutoScalingGroupRecommendationsCommand,
    ExportEBSVolumeRecommendationsCommand,
    ExportEC2InstanceRecommendationsCommand,
    ExportECSServiceRecommendationsCommand,
    ExportIdleRecommendationsCommand,
    ExportLambdaFunctionRecommendationsCommand,
    ExportLicenseRecommendationsCommand,
    ExportRDSDatabaseRecommendationsCommand,
    GetAutoScalingGroupRecommendationsCommand,
    GetEBSVolumeRecommendationsCommand,
    GetEC2InstanceRecommendationsCommand,
    GetEC2RecommendationProjectedMetricsCommand,
    GetECSServiceRecommendationProjectedMetricsCommand,
    GetECSServiceRecommendationsCommand,
    GetEffectiveRecommendationPreferencesCommand,
    GetEnrollmentStatusCommand,
    GetEnrollmentStatusesForOrganizationCommand,
    GetIdleRecommendationsCommand,
    GetLambdaFunctionRecommendationsCommand,
    GetLicenseRecommendationsCommand,
    GetRDSDatabaseRecommendationProjectedMetricsCommand,
    GetRDSDatabaseRecommendationsCommand,
    GetRecommendationPreferencesCommand,
    GetRecommendationSummariesCommand,
    PutRecommendationPreferencesCommand,
    UpdateEnrollmentStatusCommand,
};
const paginators = {
    paginateDescribeRecommendationExportJobs,
    paginateGetEnrollmentStatusesForOrganization,
    paginateGetLambdaFunctionRecommendations,
    paginateGetRecommendationPreferences,
    paginateGetRecommendationSummaries,
};
class ComputeOptimizer extends ComputeOptimizerClient {
}
smithyClient.createAggregatedClient(commands, ComputeOptimizer, { paginators });

const Status = {
    ACTIVE: "Active",
    FAILED: "Failed",
    INACTIVE: "Inactive",
    PENDING: "Pending",
};
const AllocationStrategy = {
    LOWEST_PRICE: "LowestPrice",
    PRIORITIZED: "Prioritized",
};
const AsgType = {
    MIXED_INSTANCE_TYPE: "MixedInstanceTypes",
    SINGLE_INSTANCE_TYPE: "SingleInstanceType",
};
const AutoScalingConfiguration = {
    TARGET_TRACKING_SCALING_CPU: "TargetTrackingScalingCpu",
    TARGET_TRACKING_SCALING_MEMORY: "TargetTrackingScalingMemory",
};
const Currency = {
    CNY: "CNY",
    USD: "USD",
};
const CurrentPerformanceRisk = {
    HIGH: "High",
    LOW: "Low",
    MEDIUM: "Medium",
    VERY_LOW: "VeryLow",
};
const CpuVendorArchitecture = {
    AWS_ARM64: "AWS_ARM64",
    CURRENT: "CURRENT",
};
const EnhancedInfrastructureMetrics = {
    ACTIVE: "Active",
    INACTIVE: "Inactive",
};
const ExternalMetricsSource = {
    DATADOG: "Datadog",
    DYNATRACE: "Dynatrace",
    INSTANA: "Instana",
    NEWRELIC: "NewRelic",
};
const InferredWorkloadTypesPreference = {
    ACTIVE: "Active",
    INACTIVE: "Inactive",
};
const LookBackPeriodPreference = {
    DAYS_14: "DAYS_14",
    DAYS_32: "DAYS_32",
    DAYS_93: "DAYS_93",
};
const PreferredResourceName = {
    EC2_INSTANCE_TYPES: "Ec2InstanceTypes",
};
const InstanceSavingsEstimationModeSource = {
    COST_EXPLORER_RIGHTSIZING: "CostExplorerRightsizing",
    COST_OPTIMIZATION_HUB: "CostOptimizationHub",
    PUBLIC_PRICING: "PublicPricing",
};
const CustomizableMetricName = {
    CPU_UTILIZATION: "CpuUtilization",
    MEMORY_UTILIZATION: "MemoryUtilization",
};
const CustomizableMetricHeadroom = {
    PERCENT_0: "PERCENT_0",
    PERCENT_10: "PERCENT_10",
    PERCENT_20: "PERCENT_20",
    PERCENT_30: "PERCENT_30",
};
const CustomizableMetricThreshold = {
    P90: "P90",
    P95: "P95",
    P99_5: "P99_5",
};
const Finding = {
    NOT_OPTIMIZED: "NotOptimized",
    OPTIMIZED: "Optimized",
    OVER_PROVISIONED: "Overprovisioned",
    UNDER_PROVISIONED: "Underprovisioned",
};
const InferredWorkloadType = {
    AMAZON_EMR: "AmazonEmr",
    APACHE_CASSANDRA: "ApacheCassandra",
    APACHE_HADOOP: "ApacheHadoop",
    KAFKA: "Kafka",
    MEMCACHED: "Memcached",
    NGINX: "Nginx",
    POSTGRE_SQL: "PostgreSql",
    REDIS: "Redis",
    SQLSERVER: "SQLServer",
};
const MigrationEffort = {
    HIGH: "High",
    LOW: "Low",
    MEDIUM: "Medium",
    VERY_LOW: "VeryLow",
};
const MetricName = {
    CPU: "Cpu",
    DISK_READ_BYTES_PER_SECOND: "DISK_READ_BYTES_PER_SECOND",
    DISK_READ_OPS_PER_SECOND: "DISK_READ_OPS_PER_SECOND",
    DISK_WRITE_BYTES_PER_SECOND: "DISK_WRITE_BYTES_PER_SECOND",
    DISK_WRITE_OPS_PER_SECOND: "DISK_WRITE_OPS_PER_SECOND",
    EBS_READ_BYTES_PER_SECOND: "EBS_READ_BYTES_PER_SECOND",
    EBS_READ_OPS_PER_SECOND: "EBS_READ_OPS_PER_SECOND",
    EBS_WRITE_BYTES_PER_SECOND: "EBS_WRITE_BYTES_PER_SECOND",
    EBS_WRITE_OPS_PER_SECOND: "EBS_WRITE_OPS_PER_SECOND",
    GPU_MEMORY_PERCENTAGE: "GPU_MEMORY_PERCENTAGE",
    GPU_PERCENTAGE: "GPU_PERCENTAGE",
    MEMORY: "Memory",
    NETWORK_IN_BYTES_PER_SECOND: "NETWORK_IN_BYTES_PER_SECOND",
    NETWORK_OUT_BYTES_PER_SECOND: "NETWORK_OUT_BYTES_PER_SECOND",
    NETWORK_PACKETS_IN_PER_SECOND: "NETWORK_PACKETS_IN_PER_SECOND",
    NETWORK_PACKETS_OUT_PER_SECOND: "NETWORK_PACKETS_OUT_PER_SECOND",
};
const MetricStatistic = {
    AVERAGE: "Average",
    MAXIMUM: "Maximum",
};
const RecommendationPreferenceName = {
    ENHANCED_INFRASTRUCTURE_METRICS: "EnhancedInfrastructureMetrics",
    EXTERNAL_METRICS_PREFERENCE: "ExternalMetricsPreference",
    INFERRED_WORKLOAD_TYPES: "InferredWorkloadTypes",
    LOOKBACK_PERIOD_PREFERENCE: "LookBackPeriodPreference",
    PREFERRED_RESOURCES: "PreferredResources",
    UTILIZATION_PREFERENCES: "UtilizationPreferences",
};
const ResourceType = {
    AURORA_DB_CLUSTER_STORAGE: "AuroraDBClusterStorage",
    AUTO_SCALING_GROUP: "AutoScalingGroup",
    EBS_VOLUME: "EbsVolume",
    EC2_INSTANCE: "Ec2Instance",
    ECS_SERVICE: "EcsService",
    IDLE: "Idle",
    LAMBDA_FUNCTION: "LambdaFunction",
    LICENSE: "License",
    NOT_APPLICABLE: "NotApplicable",
    RDS_DB_INSTANCE: "RdsDBInstance",
};
const ScopeName = {
    ACCOUNT_ID: "AccountId",
    ORGANIZATION: "Organization",
    RESOURCE_ARN: "ResourceArn",
};
const JobFilterName = {
    JOB_STATUS: "JobStatus",
    RESOURCE_TYPE: "ResourceType",
};
const JobStatus = {
    COMPLETE: "Complete",
    FAILED: "Failed",
    IN_PROGRESS: "InProgress",
    QUEUED: "Queued",
};
const ExportableAutoScalingGroupField = {
    ACCOUNT_ID: "AccountId",
    AUTO_SCALING_GROUP_ARN: "AutoScalingGroupArn",
    AUTO_SCALING_GROUP_NAME: "AutoScalingGroupName",
    CURRENT_CONFIGURATION_ALLOCATION_STRATEGY: "CurrentConfigurationAllocationStrategy",
    CURRENT_CONFIGURATION_DESIRED_CAPACITY: "CurrentConfigurationDesiredCapacity",
    CURRENT_CONFIGURATION_INSTANCE_TYPE: "CurrentConfigurationInstanceType",
    CURRENT_CONFIGURATION_MAX_SIZE: "CurrentConfigurationMaxSize",
    CURRENT_CONFIGURATION_MIN_SIZE: "CurrentConfigurationMinSize",
    CURRENT_CONFIGURATION_MIXED_INSTANCE_TYPES: "CurrentConfigurationMixedInstanceTypes",
    CURRENT_CONFIGURATION_TYPE: "CurrentConfigurationType",
    CURRENT_INSTANCE_GPU_INFO: "CurrentInstanceGpuInfo",
    CURRENT_MEMORY: "CurrentMemory",
    CURRENT_NETWORK: "CurrentNetwork",
    CURRENT_ON_DEMAND_PRICE: "CurrentOnDemandPrice",
    CURRENT_PERFORMANCE_RISK: "CurrentPerformanceRisk",
    CURRENT_STANDARD_ONE_YEAR_NO_UPFRONT_RESERVED_PRICE: "CurrentStandardOneYearNoUpfrontReservedPrice",
    CURRENT_STANDARD_THREE_YEAR_NO_UPFRONT_RESERVED_PRICE: "CurrentStandardThreeYearNoUpfrontReservedPrice",
    CURRENT_STORAGE: "CurrentStorage",
    CURRENT_VCPUS: "CurrentVCpus",
    EFFECTIVE_RECOMMENDATION_PREFERENCES_CPU_VENDOR_ARCHITECTURES: "EffectiveRecommendationPreferencesCpuVendorArchitectures",
    EFFECTIVE_RECOMMENDATION_PREFERENCES_ENHANCED_INFRASTRUCTURE_METRICS: "EffectiveRecommendationPreferencesEnhancedInfrastructureMetrics",
    EFFECTIVE_RECOMMENDATION_PREFERENCES_INFERRED_WORKLOAD_TYPES: "EffectiveRecommendationPreferencesInferredWorkloadTypes",
    EFFECTIVE_RECOMMENDATION_PREFERENCES_LOOKBACK_PERIOD: "EffectiveRecommendationPreferencesLookBackPeriod",
    EFFECTIVE_RECOMMENDATION_PREFERENCES_PREFERRED_RESOURCES: "EffectiveRecommendationPreferencesPreferredResources",
    EFFECTIVE_RECOMMENDATION_PREFERENCES_SAVINGS_ESTIMATION_MODE: "EffectiveRecommendationPreferencesSavingsEstimationMode",
    FINDING: "Finding",
    INFERRED_WORKLOAD_TYPES: "InferredWorkloadTypes",
    LAST_REFRESH_TIMESTAMP: "LastRefreshTimestamp",
    LOOKBACK_PERIOD_IN_DAYS: "LookbackPeriodInDays",
    RECOMMENDATION_OPTIONS_CONFIGURATION_ALLOCATION_STRATEGY: "RecommendationOptionsConfigurationAllocationStrategy",
    RECOMMENDATION_OPTIONS_CONFIGURATION_DESIRED_CAPACITY: "RecommendationOptionsConfigurationDesiredCapacity",
    RECOMMENDATION_OPTIONS_CONFIGURATION_ESTIMATED_INSTANCE_HOUR_REDUCTION_PERCENTAGE: "RecommendationOptionsConfigurationEstimatedInstanceHourReductionPercentage",
    RECOMMENDATION_OPTIONS_CONFIGURATION_INSTANCE_TYPE: "RecommendationOptionsConfigurationInstanceType",
    RECOMMENDATION_OPTIONS_CONFIGURATION_MAX_SIZE: "RecommendationOptionsConfigurationMaxSize",
    RECOMMENDATION_OPTIONS_CONFIGURATION_MIN_SIZE: "RecommendationOptionsConfigurationMinSize",
    RECOMMENDATION_OPTIONS_CONFIGURATION_MIXED_INSTANCE_TYPES: "RecommendationOptionsConfigurationMixedInstanceTypes",
    RECOMMENDATION_OPTIONS_CONFIGURATION_TYPE: "RecommendationOptionsConfigurationType",
    RECOMMENDATION_OPTIONS_ESTIMATED_MONTHLY_SAVINGS_CURRENCY: "RecommendationOptionsEstimatedMonthlySavingsCurrency",
    RECOMMENDATION_OPTIONS_ESTIMATED_MONTHLY_SAVINGS_CURRENCY_AFTER_DISCOUNTS: "RecommendationOptionsEstimatedMonthlySavingsCurrencyAfterDiscounts",
    RECOMMENDATION_OPTIONS_ESTIMATED_MONTHLY_SAVINGS_VALUE: "RecommendationOptionsEstimatedMonthlySavingsValue",
    RECOMMENDATION_OPTIONS_ESTIMATED_MONTHLY_SAVINGS_VALUE_AFTER_DISCOUNTS: "RecommendationOptionsEstimatedMonthlySavingsValueAfterDiscounts",
    RECOMMENDATION_OPTIONS_INSTANCE_GPU_INFO: "RecommendationOptionsInstanceGpuInfo",
    RECOMMENDATION_OPTIONS_MEMORY: "RecommendationOptionsMemory",
    RECOMMENDATION_OPTIONS_MIGRATION_EFFORT: "RecommendationOptionsMigrationEffort",
    RECOMMENDATION_OPTIONS_NETWORK: "RecommendationOptionsNetwork",
    RECOMMENDATION_OPTIONS_ON_DEMAND_PRICE: "RecommendationOptionsOnDemandPrice",
    RECOMMENDATION_OPTIONS_PERFORMANCE_RISK: "RecommendationOptionsPerformanceRisk",
    RECOMMENDATION_OPTIONS_PROJECTED_UTILIZATION_METRICS_CPU_MAXIMUM: "RecommendationOptionsProjectedUtilizationMetricsCpuMaximum",
    RECOMMENDATION_OPTIONS_PROJECTED_UTILIZATION_METRICS_GPU_MAXIMUM: "RecommendationOptionsProjectedUtilizationMetricsGpuPercentageMaximum",
    RECOMMENDATION_OPTIONS_PROJECTED_UTILIZATION_METRICS_GPU_MEMORY_MAXIMUM: "RecommendationOptionsProjectedUtilizationMetricsGpuMemoryPercentageMaximum",
    RECOMMENDATION_OPTIONS_PROJECTED_UTILIZATION_METRICS_MEMORY_MAXIMUM: "RecommendationOptionsProjectedUtilizationMetricsMemoryMaximum",
    RECOMMENDATION_OPTIONS_SAVINGS_OPPORTUNITY_AFTER_DISCOUNTS_PERCENTAGE: "RecommendationOptionsSavingsOpportunityAfterDiscountsPercentage",
    RECOMMENDATION_OPTIONS_SAVINGS_OPPORTUNITY_PERCENTAGE: "RecommendationOptionsSavingsOpportunityPercentage",
    RECOMMENDATION_OPTIONS_STANDARD_ONE_YEAR_NO_UPFRONT_RESERVED_PRICE: "RecommendationOptionsStandardOneYearNoUpfrontReservedPrice",
    RECOMMENDATION_OPTIONS_STANDARD_THREE_YEAR_NO_UPFRONT_RESERVED_PRICE: "RecommendationOptionsStandardThreeYearNoUpfrontReservedPrice",
    RECOMMENDATION_OPTIONS_STORAGE: "RecommendationOptionsStorage",
    RECOMMENDATION_OPTIONS_VCPUS: "RecommendationOptionsVcpus",
    UTILIZATION_METRICS_CPU_MAXIMUM: "UtilizationMetricsCpuMaximum",
    UTILIZATION_METRICS_DISK_READ_BYTES_PER_SECOND_MAXIMUM: "UtilizationMetricsDiskReadBytesPerSecondMaximum",
    UTILIZATION_METRICS_DISK_READ_OPS_PER_SECOND_MAXIMUM: "UtilizationMetricsDiskReadOpsPerSecondMaximum",
    UTILIZATION_METRICS_DISK_WRITE_BYTES_PER_SECOND_MAXIMUM: "UtilizationMetricsDiskWriteBytesPerSecondMaximum",
    UTILIZATION_METRICS_DISK_WRITE_OPS_PER_SECOND_MAXIMUM: "UtilizationMetricsDiskWriteOpsPerSecondMaximum",
    UTILIZATION_METRICS_EBS_READ_BYTES_PER_SECOND_MAXIMUM: "UtilizationMetricsEbsReadBytesPerSecondMaximum",
    UTILIZATION_METRICS_EBS_READ_OPS_PER_SECOND_MAXIMUM: "UtilizationMetricsEbsReadOpsPerSecondMaximum",
    UTILIZATION_METRICS_EBS_WRITE_BYTES_PER_SECOND_MAXIMUM: "UtilizationMetricsEbsWriteBytesPerSecondMaximum",
    UTILIZATION_METRICS_EBS_WRITE_OPS_PER_SECOND_MAXIMUM: "UtilizationMetricsEbsWriteOpsPerSecondMaximum",
    UTILIZATION_METRICS_GPU_MEMORY_PERCENTAGE_MAXIMUM: "UtilizationMetricsGpuMemoryPercentageMaximum",
    UTILIZATION_METRICS_GPU_PERCENTAGE_MAXIMUM: "UtilizationMetricsGpuPercentageMaximum",
    UTILIZATION_METRICS_MEMORY_MAXIMUM: "UtilizationMetricsMemoryMaximum",
    UTILIZATION_METRICS_NETWORK_IN_BYTES_PER_SECOND_MAXIMUM: "UtilizationMetricsNetworkInBytesPerSecondMaximum",
    UTILIZATION_METRICS_NETWORK_OUT_BYTES_PER_SECOND_MAXIMUM: "UtilizationMetricsNetworkOutBytesPerSecondMaximum",
    UTILIZATION_METRICS_NETWORK_PACKETS_IN_PER_SECOND_MAXIMUM: "UtilizationMetricsNetworkPacketsInPerSecondMaximum",
    UTILIZATION_METRICS_NETWORK_PACKETS_OUT_PER_SECOND_MAXIMUM: "UtilizationMetricsNetworkPacketsOutPerSecondMaximum",
};
const FileFormat = {
    CSV: "Csv",
};
const FilterName = {
    FINDING: "Finding",
    FINDING_REASON_CODES: "FindingReasonCodes",
    INFERRED_WORKLOAD_TYPES: "InferredWorkloadTypes",
    RECOMMENDATION_SOURCE_TYPE: "RecommendationSourceType",
};
const ExportableVolumeField = {
    ACCOUNT_ID: "AccountId",
    CURRENT_CONFIGURATION_ROOT_VOLUME: "CurrentConfigurationRootVolume",
    CURRENT_CONFIGURATION_VOLUME_BASELINE_IOPS: "CurrentConfigurationVolumeBaselineIOPS",
    CURRENT_CONFIGURATION_VOLUME_BASELINE_THROUGHPUT: "CurrentConfigurationVolumeBaselineThroughput",
    CURRENT_CONFIGURATION_VOLUME_BURST_IOPS: "CurrentConfigurationVolumeBurstIOPS",
    CURRENT_CONFIGURATION_VOLUME_BURST_THROUGHPUT: "CurrentConfigurationVolumeBurstThroughput",
    CURRENT_CONFIGURATION_VOLUME_SIZE: "CurrentConfigurationVolumeSize",
    CURRENT_CONFIGURATION_VOLUME_TYPE: "CurrentConfigurationVolumeType",
    CURRENT_MONTHLY_PRICE: "CurrentMonthlyPrice",
    CURRENT_PERFORMANCE_RISK: "CurrentPerformanceRisk",
    EFFECTIVE_RECOMMENDATION_PREFERENCES_SAVINGS_ESTIMATION_MODE: "EffectiveRecommendationPreferencesSavingsEstimationMode",
    FINDING: "Finding",
    LAST_REFRESH_TIMESTAMP: "LastRefreshTimestamp",
    LOOKBACK_PERIOD_IN_DAYS: "LookbackPeriodInDays",
    RECOMMENDATION_OPTIONS_CONFIGURATION_VOLUME_BASELINE_IOPS: "RecommendationOptionsConfigurationVolumeBaselineIOPS",
    RECOMMENDATION_OPTIONS_CONFIGURATION_VOLUME_BASELINE_THROUGHPUT: "RecommendationOptionsConfigurationVolumeBaselineThroughput",
    RECOMMENDATION_OPTIONS_CONFIGURATION_VOLUME_BURST_IOPS: "RecommendationOptionsConfigurationVolumeBurstIOPS",
    RECOMMENDATION_OPTIONS_CONFIGURATION_VOLUME_BURST_THROUGHPUT: "RecommendationOptionsConfigurationVolumeBurstThroughput",
    RECOMMENDATION_OPTIONS_CONFIGURATION_VOLUME_SIZE: "RecommendationOptionsConfigurationVolumeSize",
    RECOMMENDATION_OPTIONS_CONFIGURATION_VOLUME_TYPE: "RecommendationOptionsConfigurationVolumeType",
    RECOMMENDATION_OPTIONS_ESTIMATED_MONTHLY_SAVINGS_CURRENCY: "RecommendationOptionsEstimatedMonthlySavingsCurrency",
    RECOMMENDATION_OPTIONS_ESTIMATED_MONTHLY_SAVINGS_CURRENCY_AFTER_DISCOUNTS: "RecommendationOptionsEstimatedMonthlySavingsCurrencyAfterDiscounts",
    RECOMMENDATION_OPTIONS_ESTIMATED_MONTHLY_SAVINGS_VALUE: "RecommendationOptionsEstimatedMonthlySavingsValue",
    RECOMMENDATION_OPTIONS_ESTIMATED_MONTHLY_SAVINGS_VALUE_AFTER_DISCOUNTS: "RecommendationOptionsEstimatedMonthlySavingsValueAfterDiscounts",
    RECOMMENDATION_OPTIONS_MONTHLY_PRICE: "RecommendationOptionsMonthlyPrice",
    RECOMMENDATION_OPTIONS_PERFORMANCE_RISK: "RecommendationOptionsPerformanceRisk",
    RECOMMENDATION_OPTIONS_SAVINGS_OPPORTUNITY_AFTER_DISCOUNTS_PERCENTAGE: "RecommendationOptionsSavingsOpportunityAfterDiscountsPercentage",
    RECOMMENDATION_OPTIONS_SAVINGS_OPPORTUNITY_PERCENTAGE: "RecommendationOptionsSavingsOpportunityPercentage",
    ROOT_VOLUME: "RootVolume",
    TAGS: "Tags",
    UTILIZATION_METRICS_VOLUME_READ_BYTES_PER_SECOND_MAXIMUM: "UtilizationMetricsVolumeReadBytesPerSecondMaximum",
    UTILIZATION_METRICS_VOLUME_READ_OPS_PER_SECOND_MAXIMUM: "UtilizationMetricsVolumeReadOpsPerSecondMaximum",
    UTILIZATION_METRICS_VOLUME_WRITE_BYTES_PER_SECOND_MAXIMUM: "UtilizationMetricsVolumeWriteBytesPerSecondMaximum",
    UTILIZATION_METRICS_VOLUME_WRITE_OPS_PER_SECOND_MAXIMUM: "UtilizationMetricsVolumeWriteOpsPerSecondMaximum",
    VOLUME_ARN: "VolumeArn",
};
const EBSFilterName = {
    FINDING: "Finding",
};
const ExportableInstanceField = {
    ACCOUNT_ID: "AccountId",
    CURRENT_INSTANCE_GPU_INFO: "CurrentInstanceGpuInfo",
    CURRENT_INSTANCE_TYPE: "CurrentInstanceType",
    CURRENT_MEMORY: "CurrentMemory",
    CURRENT_NETWORK: "CurrentNetwork",
    CURRENT_ON_DEMAND_PRICE: "CurrentOnDemandPrice",
    CURRENT_PERFORMANCE_RISK: "CurrentPerformanceRisk",
    CURRENT_STANDARD_ONE_YEAR_NO_UPFRONT_RESERVED_PRICE: "CurrentStandardOneYearNoUpfrontReservedPrice",
    CURRENT_STANDARD_THREE_YEAR_NO_UPFRONT_RESERVED_PRICE: "CurrentStandardThreeYearNoUpfrontReservedPrice",
    CURRENT_STORAGE: "CurrentStorage",
    CURRENT_VCPUS: "CurrentVCpus",
    EFFECTIVE_RECOMMENDATION_PREFERENCES_CPU_VENDOR_ARCHITECTURES: "EffectiveRecommendationPreferencesCpuVendorArchitectures",
    EFFECTIVE_RECOMMENDATION_PREFERENCES_ENHANCED_INFRASTRUCTURE_METRICS: "EffectiveRecommendationPreferencesEnhancedInfrastructureMetrics",
    EFFECTIVE_RECOMMENDATION_PREFERENCES_EXTERNAL_METRICS_SOURCE: "EffectiveRecommendationPreferencesExternalMetricsSource",
    EFFECTIVE_RECOMMENDATION_PREFERENCES_INFERRED_WORKLOAD_TYPES: "EffectiveRecommendationPreferencesInferredWorkloadTypes",
    EFFECTIVE_RECOMMENDATION_PREFERENCES_LOOKBACK_PERIOD: "EffectiveRecommendationPreferencesLookBackPeriod",
    EFFECTIVE_RECOMMENDATION_PREFERENCES_PREFERRED_RESOURCES: "EffectiveRecommendationPreferencesPreferredResources",
    EFFECTIVE_RECOMMENDATION_PREFERENCES_SAVINGS_ESTIMATION_MODE: "EffectiveRecommendationPreferencesSavingsEstimationMode",
    EFFECTIVE_RECOMMENDATION_PREFERENCES_UTILIZATION_PREFERENCES: "EffectiveRecommendationPreferencesUtilizationPreferences",
    EXTERNAL_METRIC_STATUS_CODE: "ExternalMetricStatusCode",
    EXTERNAL_METRIC_STATUS_REASON: "ExternalMetricStatusReason",
    FINDING: "Finding",
    Finding_Reason_Codes: "FindingReasonCodes",
    IDLE: "Idle",
    INFERRED_WORKLOAD_TYPES: "InferredWorkloadTypes",
    INSTANCE_ARN: "InstanceArn",
    INSTANCE_NAME: "InstanceName",
    INSTANCE_STATE: "InstanceState",
    LAST_REFRESH_TIMESTAMP: "LastRefreshTimestamp",
    LOOKBACK_PERIOD_IN_DAYS: "LookbackPeriodInDays",
    RECOMMENDATIONS_SOURCES_RECOMMENDATION_SOURCE_ARN: "RecommendationsSourcesRecommendationSourceArn",
    RECOMMENDATIONS_SOURCES_RECOMMENDATION_SOURCE_TYPE: "RecommendationsSourcesRecommendationSourceType",
    RECOMMENDATION_OPTIONS_ESTIMATED_MONTHLY_SAVINGS_CURRENCY: "RecommendationOptionsEstimatedMonthlySavingsCurrency",
    RECOMMENDATION_OPTIONS_ESTIMATED_MONTHLY_SAVINGS_CURRENCY_AFTER_DISCOUNTS: "RecommendationOptionsEstimatedMonthlySavingsCurrencyAfterDiscounts",
    RECOMMENDATION_OPTIONS_ESTIMATED_MONTHLY_SAVINGS_VALUE: "RecommendationOptionsEstimatedMonthlySavingsValue",
    RECOMMENDATION_OPTIONS_ESTIMATED_MONTHLY_SAVINGS_VALUE_AFTER_DISCOUNTS: "RecommendationOptionsEstimatedMonthlySavingsValueAfterDiscounts",
    RECOMMENDATION_OPTIONS_INSTANCE_GPU_INFO: "RecommendationOptionsInstanceGpuInfo",
    RECOMMENDATION_OPTIONS_INSTANCE_TYPE: "RecommendationOptionsInstanceType",
    RECOMMENDATION_OPTIONS_MEMORY: "RecommendationOptionsMemory",
    RECOMMENDATION_OPTIONS_MIGRATION_EFFORT: "RecommendationOptionsMigrationEffort",
    RECOMMENDATION_OPTIONS_NETWORK: "RecommendationOptionsNetwork",
    RECOMMENDATION_OPTIONS_ON_DEMAND_PRICE: "RecommendationOptionsOnDemandPrice",
    RECOMMENDATION_OPTIONS_PERFORMANCE_RISK: "RecommendationOptionsPerformanceRisk",
    RECOMMENDATION_OPTIONS_PLATFORM_DIFFERENCES: "RecommendationOptionsPlatformDifferences",
    RECOMMENDATION_OPTIONS_PROJECTED_UTILIZATION_METRICS_CPU_MAXIMUM: "RecommendationOptionsProjectedUtilizationMetricsCpuMaximum",
    RECOMMENDATION_OPTIONS_PROJECTED_UTILIZATION_METRICS_GPU_MEMORY_PERCENTAGE_MAXIMUM: "RecommendationOptionsProjectedUtilizationMetricsGpuMemoryPercentageMaximum",
    RECOMMENDATION_OPTIONS_PROJECTED_UTILIZATION_METRICS_GPU_PERCENTAGE_MAXIMUM: "RecommendationOptionsProjectedUtilizationMetricsGpuPercentageMaximum",
    RECOMMENDATION_OPTIONS_PROJECTED_UTILIZATION_METRICS_MEMORY_MAXIMUM: "RecommendationOptionsProjectedUtilizationMetricsMemoryMaximum",
    RECOMMENDATION_OPTIONS_SAVINGS_OPPORTUNITY_AFTER_DISCOUNTS_PERCENTAGE: "RecommendationOptionsSavingsOpportunityAfterDiscountsPercentage",
    RECOMMENDATION_OPTIONS_SAVINGS_OPPORTUNITY_PERCENTAGE: "RecommendationOptionsSavingsOpportunityPercentage",
    RECOMMENDATION_OPTIONS_STANDARD_ONE_YEAR_NO_UPFRONT_RESERVED_PRICE: "RecommendationOptionsStandardOneYearNoUpfrontReservedPrice",
    RECOMMENDATION_OPTIONS_STANDARD_THREE_YEAR_NO_UPFRONT_RESERVED_PRICE: "RecommendationOptionsStandardThreeYearNoUpfrontReservedPrice",
    RECOMMENDATION_OPTIONS_STORAGE: "RecommendationOptionsStorage",
    RECOMMENDATION_OPTIONS_VCPUS: "RecommendationOptionsVcpus",
    TAGS: "Tags",
    UTILIZATION_METRICS_CPU_MAXIMUM: "UtilizationMetricsCpuMaximum",
    UTILIZATION_METRICS_DISK_READ_BYTES_PER_SECOND_MAXIMUM: "UtilizationMetricsDiskReadBytesPerSecondMaximum",
    UTILIZATION_METRICS_DISK_READ_OPS_PER_SECOND_MAXIMUM: "UtilizationMetricsDiskReadOpsPerSecondMaximum",
    UTILIZATION_METRICS_DISK_WRITE_BYTES_PER_SECOND_MAXIMUM: "UtilizationMetricsDiskWriteBytesPerSecondMaximum",
    UTILIZATION_METRICS_DISK_WRITE_OPS_PER_SECOND_MAXIMUM: "UtilizationMetricsDiskWriteOpsPerSecondMaximum",
    UTILIZATION_METRICS_EBS_READ_BYTES_PER_SECOND_MAXIMUM: "UtilizationMetricsEbsReadBytesPerSecondMaximum",
    UTILIZATION_METRICS_EBS_READ_OPS_PER_SECOND_MAXIMUM: "UtilizationMetricsEbsReadOpsPerSecondMaximum",
    UTILIZATION_METRICS_EBS_WRITE_BYTES_PER_SECOND_MAXIMUM: "UtilizationMetricsEbsWriteBytesPerSecondMaximum",
    UTILIZATION_METRICS_EBS_WRITE_OPS_PER_SECOND_MAXIMUM: "UtilizationMetricsEbsWriteOpsPerSecondMaximum",
    UTILIZATION_METRICS_GPU_MEMORY_PERCENTAGE_MAXIMUM: "UtilizationMetricsGpuMemoryPercentageMaximum",
    UTILIZATION_METRICS_GPU_PERCENTAGE_MAXIMUM: "UtilizationMetricsGpuPercentageMaximum",
    UTILIZATION_METRICS_MEMORY_MAXIMUM: "UtilizationMetricsMemoryMaximum",
    UTILIZATION_METRICS_NETWORK_IN_BYTES_PER_SECOND_MAXIMUM: "UtilizationMetricsNetworkInBytesPerSecondMaximum",
    UTILIZATION_METRICS_NETWORK_OUT_BYTES_PER_SECOND_MAXIMUM: "UtilizationMetricsNetworkOutBytesPerSecondMaximum",
    UTILIZATION_METRICS_NETWORK_PACKETS_IN_PER_SECOND_MAXIMUM: "UtilizationMetricsNetworkPacketsInPerSecondMaximum",
    UTILIZATION_METRICS_NETWORK_PACKETS_OUT_PER_SECOND_MAXIMUM: "UtilizationMetricsNetworkPacketsOutPerSecondMaximum",
};
const ExportableECSServiceField = {
    ACCOUNT_ID: "AccountId",
    CURRENT_PERFORMANCE_RISK: "CurrentPerformanceRisk",
    CURRENT_SERVICE_CONFIGURATION_AUTO_SCALING_CONFIGURATION: "CurrentServiceConfigurationAutoScalingConfiguration",
    CURRENT_SERVICE_CONFIGURATION_CPU: "CurrentServiceConfigurationCpu",
    CURRENT_SERVICE_CONFIGURATION_MEMORY: "CurrentServiceConfigurationMemory",
    CURRENT_SERVICE_CONFIGURATION_TASK_DEFINITION_ARN: "CurrentServiceConfigurationTaskDefinitionArn",
    CURRENT_SERVICE_CONTAINER_CONFIGURATIONS: "CurrentServiceContainerConfigurations",
    EFFECTIVE_RECOMMENDATION_PREFERENCES_SAVINGS_ESTIMATION_MODE: "EffectiveRecommendationPreferencesSavingsEstimationMode",
    FINDING: "Finding",
    FINDING_REASON_CODES: "FindingReasonCodes",
    LAST_REFRESH_TIMESTAMP: "LastRefreshTimestamp",
    LAUNCH_TYPE: "LaunchType",
    LOOKBACK_PERIOD_IN_DAYS: "LookbackPeriodInDays",
    RECOMMENDATION_OPTIONS_CONTAINER_RECOMMENDATIONS: "RecommendationOptionsContainerRecommendations",
    RECOMMENDATION_OPTIONS_CPU: "RecommendationOptionsCpu",
    RECOMMENDATION_OPTIONS_ESTIMATED_MONTHLY_SAVINGS_CURRENCY: "RecommendationOptionsEstimatedMonthlySavingsCurrency",
    RECOMMENDATION_OPTIONS_ESTIMATED_MONTHLY_SAVINGS_CURRENCY_AFTER_DISCOUNTS: "RecommendationOptionsEstimatedMonthlySavingsCurrencyAfterDiscounts",
    RECOMMENDATION_OPTIONS_ESTIMATED_MONTHLY_SAVINGS_VALUE: "RecommendationOptionsEstimatedMonthlySavingsValue",
    RECOMMENDATION_OPTIONS_ESTIMATED_MONTHLY_SAVINGS_VALUE_AFTER_DISCOUNTS: "RecommendationOptionsEstimatedMonthlySavingsValueAfterDiscounts",
    RECOMMENDATION_OPTIONS_MEMORY: "RecommendationOptionsMemory",
    RECOMMENDATION_OPTIONS_PROJECTED_UTILIZATION_METRICS_CPU_MAXIMUM: "RecommendationOptionsProjectedUtilizationMetricsCpuMaximum",
    RECOMMENDATION_OPTIONS_PROJECTED_UTILIZATION_METRICS_MEMORY_MAXIMUM: "RecommendationOptionsProjectedUtilizationMetricsMemoryMaximum",
    RECOMMENDATION_OPTIONS_SAVINGS_OPPORTUNITY_AFTER_DISCOUNTS_PERCENTAGE: "RecommendationOptionsSavingsOpportunityAfterDiscountsPercentage",
    RECOMMENDATION_OPTIONS_SAVINGS_OPPORTUNITY_PERCENTAGE: "RecommendationOptionsSavingsOpportunityPercentage",
    SERVICE_ARN: "ServiceArn",
    TAGS: "Tags",
    UTILIZATION_METRICS_CPU_MAXIMUM: "UtilizationMetricsCpuMaximum",
    UTILIZATION_METRICS_MEMORY_MAXIMUM: "UtilizationMetricsMemoryMaximum",
};
const ECSServiceRecommendationFilterName = {
    FINDING: "Finding",
    FINDING_REASON_CODE: "FindingReasonCode",
};
const ExportableIdleField = {
    ACCOUNT_ID: "AccountId",
    FINDING: "Finding",
    FINDING_DESCRIPTION: "FindingDescription",
    LAST_REFRESH_TIMESTAMP: "LastRefreshTimestamp",
    LOOKBACK_PERIOD_IN_DAYS: "LookbackPeriodInDays",
    RESOURCE_ARN: "ResourceArn",
    RESOURCE_ID: "ResourceId",
    RESOURCE_TYPE: "ResourceType",
    SAVINGS_OPPORTUNITY: "SavingsOpportunity",
    SAVINGS_OPPORTUNITY_AFTER_DISCOUNT: "SavingsOpportunityAfterDiscount",
    TAGS: "Tags",
    UTILIZATION_METRICS_ACTIVE_CONNECTION_COUNT_MAXIMUM: "UtilizationMetricsActiveConnectionCountMaximum",
    UTILIZATION_METRICS_CPU_MAXIMUM: "UtilizationMetricsCpuMaximum",
    UTILIZATION_METRICS_DATABASE_CONNECTIONS_MAXIMUM: "UtilizationMetricsDatabaseConnectionsMaximum",
    UTILIZATION_METRICS_EBS_VOLUME_READ_IOPS_MAXIMUM: "UtilizationMetricsEBSVolumeReadIOPSMaximum",
    UTILIZATION_METRICS_EBS_VOLUME_WRITE_IOPS_MAXIMUM: "UtilizationMetricsEBSVolumeWriteIOPSMaximum",
    UTILIZATION_METRICS_MEMORY_MAXIMUM: "UtilizationMetricsMemoryMaximum",
    UTILIZATION_METRICS_NETWORK_IN_BYTES_PER_SECOND_MAXIMUM: "UtilizationMetricsNetworkInBytesPerSecondMaximum",
    UTILIZATION_METRICS_NETWORK_OUT_BYTES_PER_SECOND_MAXIMUM: "UtilizationMetricsNetworkOutBytesPerSecondMaximum",
    UTILIZATION_METRICS_PACKETS_IN_FROM_DESTINATION_MAXIMUM: "UtilizationMetricsPacketsInFromDestinationMaximum",
    UTILIZATION_METRICS_PACKETS_IN_FROM_SOURCE_MAXIMUM: "UtilizationMetricsPacketsInFromSourceMaximum",
    UTILIZATION_METRICS_VOLUME_READ_OPS_PER_SECOND_MAXIMUM: "UtilizationMetricsVolumeReadOpsPerSecondMaximum",
    UTILIZATION_METRICS_VOLUME_WRITE_OPS_PER_SECOND_MAXIMUM: "UtilizationMetricsVolumeWriteOpsPerSecondMaximum",
};
const IdleRecommendationFilterName = {
    FINDING: "Finding",
    RESOURCE_TYPE: "ResourceType",
};
const ExportableLambdaFunctionField = {
    ACCOUNT_ID: "AccountId",
    CURRENT_CONFIGURATION_MEMORY_SIZE: "CurrentConfigurationMemorySize",
    CURRENT_CONFIGURATION_TIMEOUT: "CurrentConfigurationTimeout",
    CURRENT_COST_AVERAGE: "CurrentCostAverage",
    CURRENT_COST_TOTAL: "CurrentCostTotal",
    CURRENT_PERFORMANCE_RISK: "CurrentPerformanceRisk",
    EFFECTIVE_RECOMMENDATION_PREFERENCES_SAVINGS_ESTIMATION_MODE: "EffectiveRecommendationPreferencesSavingsEstimationMode",
    FINDING: "Finding",
    FINDING_REASON_CODES: "FindingReasonCodes",
    FUNCTION_ARN: "FunctionArn",
    FUNCTION_VERSION: "FunctionVersion",
    LAST_REFRESH_TIMESTAMP: "LastRefreshTimestamp",
    LOOKBACK_PERIOD_IN_DAYS: "LookbackPeriodInDays",
    NUMBER_OF_INVOCATIONS: "NumberOfInvocations",
    RECOMMENDATION_OPTIONS_CONFIGURATION_MEMORY_SIZE: "RecommendationOptionsConfigurationMemorySize",
    RECOMMENDATION_OPTIONS_COST_HIGH: "RecommendationOptionsCostHigh",
    RECOMMENDATION_OPTIONS_COST_LOW: "RecommendationOptionsCostLow",
    RECOMMENDATION_OPTIONS_ESTIMATED_MONTHLY_SAVINGS_CURRENCY: "RecommendationOptionsEstimatedMonthlySavingsCurrency",
    RECOMMENDATION_OPTIONS_ESTIMATED_MONTHLY_SAVINGS_CURRENCY_AFTER_DISCOUNTS: "RecommendationOptionsEstimatedMonthlySavingsCurrencyAfterDiscounts",
    RECOMMENDATION_OPTIONS_ESTIMATED_MONTHLY_SAVINGS_VALUE: "RecommendationOptionsEstimatedMonthlySavingsValue",
    RECOMMENDATION_OPTIONS_ESTIMATED_MONTHLY_SAVINGS_VALUE_AFTER_DISCOUNTS: "RecommendationOptionsEstimatedMonthlySavingsValueAfterDiscounts",
    RECOMMENDATION_OPTIONS_PROJECTED_UTILIZATION_METRICS_DURATION_EXPECTED: "RecommendationOptionsProjectedUtilizationMetricsDurationExpected",
    RECOMMENDATION_OPTIONS_PROJECTED_UTILIZATION_METRICS_DURATION_LOWER_BOUND: "RecommendationOptionsProjectedUtilizationMetricsDurationLowerBound",
    RECOMMENDATION_OPTIONS_PROJECTED_UTILIZATION_METRICS_DURATION_UPPER_BOUND: "RecommendationOptionsProjectedUtilizationMetricsDurationUpperBound",
    RECOMMENDATION_OPTIONS_SAVINGS_OPPORTUNITY_AFTER_DISCOUNTS_PERCENTAGE: "RecommendationOptionsSavingsOpportunityAfterDiscountsPercentage",
    RECOMMENDATION_OPTIONS_SAVINGS_OPPORTUNITY_PERCENTAGE: "RecommendationOptionsSavingsOpportunityPercentage",
    TAGS: "Tags",
    UTILIZATION_METRICS_DURATION_AVERAGE: "UtilizationMetricsDurationAverage",
    UTILIZATION_METRICS_DURATION_MAXIMUM: "UtilizationMetricsDurationMaximum",
    UTILIZATION_METRICS_MEMORY_AVERAGE: "UtilizationMetricsMemoryAverage",
    UTILIZATION_METRICS_MEMORY_MAXIMUM: "UtilizationMetricsMemoryMaximum",
};
const LambdaFunctionRecommendationFilterName = {
    FINDING: "Finding",
    FINDING_REASON_CODE: "FindingReasonCode",
};
const ExportableLicenseField = {
    ACCOUNT_ID: "AccountId",
    CURRENT_LICENSE_CONFIGURATION_INSTANCE_TYPE: "CurrentLicenseConfigurationInstanceType",
    CURRENT_LICENSE_CONFIGURATION_LICENSE_EDITION: "CurrentLicenseConfigurationLicenseEdition",
    CURRENT_LICENSE_CONFIGURATION_LICENSE_MODEL: "CurrentLicenseConfigurationLicenseModel",
    CURRENT_LICENSE_CONFIGURATION_LICENSE_NAME: "CurrentLicenseConfigurationLicenseName",
    CURRENT_LICENSE_CONFIGURATION_LICENSE_VERSION: "CurrentLicenseConfigurationLicenseVersion",
    CURRENT_LICENSE_CONFIGURATION_METRICS_SOURCE: "CurrentLicenseConfigurationMetricsSource",
    CURRENT_LICENSE_CONFIGURATION_NUMBER_OF_CORES: "CurrentLicenseConfigurationNumberOfCores",
    CURRENT_LICENSE_CONFIGURATION_OPERATING_SYSTEM: "CurrentLicenseConfigurationOperatingSystem",
    LAST_REFRESH_TIMESTAMP: "LastRefreshTimestamp",
    LICENSE_FINDING: "Finding",
    LICENSE_FINDING_REASON_CODES: "FindingReasonCodes",
    LOOKBACK_PERIOD_IN_DAYS: "LookbackPeriodInDays",
    RECOMMENDATION_OPTIONS_ESTIMATED_MONTHLY_SAVINGS_CURRENCY: "RecommendationOptionsEstimatedMonthlySavingsCurrency",
    RECOMMENDATION_OPTIONS_ESTIMATED_MONTHLY_SAVINGS_VALUE: "RecommendationOptionsEstimatedMonthlySavingsValue",
    RECOMMENDATION_OPTIONS_LICENSE_EDITION: "RecommendationOptionsLicenseEdition",
    RECOMMENDATION_OPTIONS_LICENSE_MODEL: "RecommendationOptionsLicenseModel",
    RECOMMENDATION_OPTIONS_OPERATING_SYSTEM: "RecommendationOptionsOperatingSystem",
    RECOMMENDATION_OPTIONS_SAVINGS_OPPORTUNITY_PERCENTAGE: "RecommendationOptionsSavingsOpportunityPercentage",
    RESOURCE_ARN: "ResourceArn",
    TAGS: "Tags",
};
const LicenseRecommendationFilterName = {
    LICENSE_FINDING: "Finding",
    LICENSE_FINDING_REASON_CODE: "FindingReasonCode",
    LICENSE_NAME: "LicenseName",
};
const ExportableRDSDBField = {
    ACCOUNT_ID: "AccountId",
    CLUSTER_WRITER: "ClusterWriter",
    CURRENT_DB_INSTANCE_CLASS: "CurrentDBInstanceClass",
    CURRENT_INSTANCE_ON_DEMAND_HOURLY_PRICE: "CurrentInstanceOnDemandHourlyPrice",
    CURRENT_INSTANCE_PERFORMANCE_RISK: "CurrentInstancePerformanceRisk",
    CURRENT_STORAGE_CONFIGURATION_ALLOCATED_STORAGE: "CurrentStorageConfigurationAllocatedStorage",
    CURRENT_STORAGE_CONFIGURATION_IOPS: "CurrentStorageConfigurationIOPS",
    CURRENT_STORAGE_CONFIGURATION_MAX_ALLOCATED_STORAGE: "CurrentStorageConfigurationMaxAllocatedStorage",
    CURRENT_STORAGE_CONFIGURATION_STORAGE_THROUGHPUT: "CurrentStorageConfigurationStorageThroughput",
    CURRENT_STORAGE_CONFIGURATION_STORAGE_TYPE: "CurrentStorageConfigurationStorageType",
    CURRENT_STORAGE_ESTIMATED_CLUSTER_INSTANCE_ON_DEMAND_MONTHLY_COST: "CurrentStorageEstimatedClusterInstanceOnDemandMonthlyCost",
    CURRENT_STORAGE_ESTIMATED_CLUSTER_STORAGE_IO_ON_DEMAND_MONTHLY_COST: "CurrentStorageEstimatedClusterStorageIOOnDemandMonthlyCost",
    CURRENT_STORAGE_ESTIMATED_CLUSTER_STORAGE_ON_DEMAND_MONTHLY_COST: "CurrentStorageEstimatedClusterStorageOnDemandMonthlyCost",
    CURRENT_STORAGE_ESTIMATED_MONTHLY_VOLUME_IOPS_COST_VARIATION: "CurrentStorageEstimatedMonthlyVolumeIOPsCostVariation",
    CURRENT_STORAGE_ON_DEMAND_MONTHLY_PRICE: "CurrentStorageOnDemandMonthlyPrice",
    DB_CLUSTER_IDENTIFIER: "DBClusterIdentifier",
    EFFECTIVE_RECOMMENDATION_PREFERENCES_CPU_VENDOR_ARCHITECTURES: "EffectiveRecommendationPreferencesCpuVendorArchitectures",
    EFFECTIVE_RECOMMENDATION_PREFERENCES_ENHANCED_INFRASTRUCTURE_METRICS: "EffectiveRecommendationPreferencesEnhancedInfrastructureMetrics",
    EFFECTIVE_RECOMMENDATION_PREFERENCES_LOOKBACK_PERIOD: "EffectiveRecommendationPreferencesLookBackPeriod",
    EFFECTIVE_RECOMMENDATION_PREFERENCES_SAVINGS_ESTIMATION_MODE: "EffectiveRecommendationPreferencesSavingsEstimationMode",
    ENGINE: "Engine",
    ENGINE_VERSION: "EngineVersion",
    IDLE: "Idle",
    INSTANCE_FINDING: "InstanceFinding",
    INSTANCE_FINDING_REASON_CODES: "InstanceFindingReasonCodes",
    INSTANCE_RECOMMENDATION_OPTIONS_DB_INSTANCE_CLASS: "InstanceRecommendationOptionsDBInstanceClass",
    INSTANCE_RECOMMENDATION_OPTIONS_ESTIMATED_MONTHLY_SAVINGS_CURRENCY: "InstanceRecommendationOptionsEstimatedMonthlySavingsCurrency",
    INSTANCE_RECOMMENDATION_OPTIONS_ESTIMATED_MONTHLY_SAVINGS_CURRENCY_AFTER_DISCOUNTS: "InstanceRecommendationOptionsEstimatedMonthlySavingsCurrencyAfterDiscounts",
    INSTANCE_RECOMMENDATION_OPTIONS_ESTIMATED_MONTHLY_SAVINGS_VALUE: "InstanceRecommendationOptionsEstimatedMonthlySavingsValue",
    INSTANCE_RECOMMENDATION_OPTIONS_ESTIMATED_MONTHLY_SAVINGS_VALUE_AFTER_DISCOUNTS: "InstanceRecommendationOptionsEstimatedMonthlySavingsValueAfterDiscounts",
    INSTANCE_RECOMMENDATION_OPTIONS_INSTANCE_ON_DEMAND_HOURLY_PRICE: "InstanceRecommendationOptionsInstanceOnDemandHourlyPrice",
    INSTANCE_RECOMMENDATION_OPTIONS_PERFORMANCE_RISK: "InstanceRecommendationOptionsPerformanceRisk",
    INSTANCE_RECOMMENDATION_OPTIONS_PROJECTED_UTILIZATION_METRICS_CPU_MAXIMUM: "InstanceRecommendationOptionsProjectedUtilizationMetricsCpuMaximum",
    INSTANCE_RECOMMENDATION_OPTIONS_RANK: "InstanceRecommendationOptionsRank",
    INSTANCE_RECOMMENDATION_OPTIONS_SAVINGS_OPPORTUNITY_AFTER_DISCOUNTS_PERCENTAGE: "InstanceRecommendationOptionsSavingsOpportunityAfterDiscountsPercentage",
    INSTANCE_RECOMMENDATION_OPTIONS_SAVINGS_OPPORTUNITY_PERCENTAGE: "InstanceRecommendationOptionsSavingsOpportunityPercentage",
    LAST_REFRESH_TIMESTAMP: "LastRefreshTimestamp",
    LOOKBACK_PERIOD_IN_DAYS: "LookbackPeriodInDays",
    MULTI_AZ_DB_INSTANCE: "MultiAZDBInstance",
    PROMOTION_TIER: "PromotionTier",
    RESOURCE_ARN: "ResourceArn",
    STORAGE_FINDING: "StorageFinding",
    STORAGE_FINDING_REASON_CODES: "StorageFindingReasonCodes",
    STORAGE_RECOMMENDATION_OPTIONS_ALLOCATED_STORAGE: "StorageRecommendationOptionsAllocatedStorage",
    STORAGE_RECOMMENDATION_OPTIONS_ESTIMATED_CLUSTER_INSTANCE_ON_DEMAND_MONTHLY_COST: "StorageRecommendationOptionsEstimatedClusterInstanceOnDemandMonthlyCost",
    STORAGE_RECOMMENDATION_OPTIONS_ESTIMATED_CLUSTER_STORAGE_IO_ON_DEMAND_MONTHLY_COST: "StorageRecommendationOptionsEstimatedClusterStorageIOOnDemandMonthlyCost",
    STORAGE_RECOMMENDATION_OPTIONS_ESTIMATED_CLUSTER_STORAGE_ON_DEMAND_MONTHLY_COST: "StorageRecommendationOptionsEstimatedClusterStorageOnDemandMonthlyCost",
    STORAGE_RECOMMENDATION_OPTIONS_ESTIMATED_MONTHLY_SAVINGS_CURRENCY: "StorageRecommendationOptionsEstimatedMonthlySavingsCurrency",
    STORAGE_RECOMMENDATION_OPTIONS_ESTIMATED_MONTHLY_SAVINGS_CURRENCY_AFTER_DISCOUNTS: "StorageRecommendationOptionsEstimatedMonthlySavingsCurrencyAfterDiscounts",
    STORAGE_RECOMMENDATION_OPTIONS_ESTIMATED_MONTHLY_SAVINGS_VALUE: "StorageRecommendationOptionsEstimatedMonthlySavingsValue",
    STORAGE_RECOMMENDATION_OPTIONS_ESTIMATED_MONTHLY_SAVINGS_VALUE_AFTER_DISCOUNTS: "StorageRecommendationOptionsEstimatedMonthlySavingsValueAfterDiscounts",
    STORAGE_RECOMMENDATION_OPTIONS_ESTIMATED_MONTHLY_VOLUME_IOPS_COST_VARIATION: "StorageRecommendationOptionsEstimatedMonthlyVolumeIOPsCostVariation",
    STORAGE_RECOMMENDATION_OPTIONS_IOPS: "StorageRecommendationOptionsIOPS",
    STORAGE_RECOMMENDATION_OPTIONS_MAX_ALLOCATED_STORAGE: "StorageRecommendationOptionsMaxAllocatedStorage",
    STORAGE_RECOMMENDATION_OPTIONS_ON_DEMAND_MONTHLY_PRICE: "StorageRecommendationOptionsOnDemandMonthlyPrice",
    STORAGE_RECOMMENDATION_OPTIONS_RANK: "StorageRecommendationOptionsRank",
    STORAGE_RECOMMENDATION_OPTIONS_SAVINGS_OPPORTUNITY_AFTER_DISCOUNTS_PERCENTAGE: "StorageRecommendationOptionsSavingsOpportunityAfterDiscountsPercentage",
    STORAGE_RECOMMENDATION_OPTIONS_SAVINGS_OPPORTUNITY_PERCENTAGE: "StorageRecommendationOptionsSavingsOpportunityPercentage",
    STORAGE_RECOMMENDATION_OPTIONS_STORAGE_THROUGHPUT: "StorageRecommendationOptionsStorageThroughput",
    STORAGE_RECOMMENDATION_OPTIONS_STORAGE_TYPE: "StorageRecommendationOptionsStorageType",
    TAGS: "Tags",
    UTILIZATION_METRICS_AURORA_MEMORY_HEALTH_STATE_MAXIMUM: "UtilizationMetricsAuroraMemoryHealthStateMaximum",
    UTILIZATION_METRICS_AURORA_MEMORY_NUM_DECLINED_SQL_TOTAL_MAXIMUM: "UtilizationMetricsAuroraMemoryNumDeclinedSqlTotalMaximum",
    UTILIZATION_METRICS_AURORA_MEMORY_NUM_KILL_CONN_TOTAL_MAXIMUM: "UtilizationMetricsAuroraMemoryNumKillConnTotalMaximum",
    UTILIZATION_METRICS_AURORA_MEMORY_NUM_KILL_QUERY_TOTAL_MAXIMUM: "UtilizationMetricsAuroraMemoryNumKillQueryTotalMaximum",
    UTILIZATION_METRICS_CPU_MAXIMUM: "UtilizationMetricsCpuMaximum",
    UTILIZATION_METRICS_DATABASE_CONNECTIONS_MAXIMUM: "UtilizationMetricsDatabaseConnectionsMaximum",
    UTILIZATION_METRICS_EBS_VOLUME_READ_IOPS_MAXIMUM: "UtilizationMetricsEBSVolumeReadIOPSMaximum",
    UTILIZATION_METRICS_EBS_VOLUME_READ_THROUGHPUT_MAXIMUM: "UtilizationMetricsEBSVolumeReadThroughputMaximum",
    UTILIZATION_METRICS_EBS_VOLUME_STORAGE_SPACE_UTILIZATION_MAXIMUM: "UtilizationMetricsEBSVolumeStorageSpaceUtilizationMaximum",
    UTILIZATION_METRICS_EBS_VOLUME_WRITE_IOPS_MAXIMUM: "UtilizationMetricsEBSVolumeWriteIOPSMaximum",
    UTILIZATION_METRICS_EBS_VOLUME_WRITE_THROUGHPUT_MAXIMUM: "UtilizationMetricsEBSVolumeWriteThroughputMaximum",
    UTILIZATION_METRICS_MEMORY_MAXIMUM: "UtilizationMetricsMemoryMaximum",
    UTILIZATION_METRICS_NETWORK_RECEIVE_THROUGHPUT_MAXIMUM: "UtilizationMetricsNetworkReceiveThroughputMaximum",
    UTILIZATION_METRICS_NETWORK_TRANSMIT_THROUGHPUT_MAXIMUM: "UtilizationMetricsNetworkTransmitThroughputMaximum",
    UTILIZATION_METRICS_READ_IOPS_EPHEMERAL_STORAGE_MAXIMUM: "UtilizationMetricsReadIOPSEphemeralStorageMaximum",
    UTILIZATION_METRICS_STORAGE_NETWORK_RECEIVE_THROUGHPUT_MAXIMUM: "UtilizationMetricsStorageNetworkReceiveThroughputMaximum",
    UTILIZATION_METRICS_STORAGE_NETWORK_TRANSMIT_THROUGHPUT_MAXIMUM: "UtilizationMetricsStorageNetworkTransmitThroughputMaximum",
    UTILIZATION_METRICS_VOLUME_BYTES_USED_AVERAGE: "UtilizationMetricsVolumeBytesUsedAverage",
    UTILIZATION_METRICS_VOLUME_READ_IOPS_AVERAGE: "UtilizationMetricsVolumeReadIOPsAverage",
    UTILIZATION_METRICS_VOLUME_WRITE_IOPS_AVERAGE: "UtilizationMetricsVolumeWriteIOPsAverage",
    UTILIZATION_METRICS_WRITE_IOPS_EPHEMERAL_STORAGE_MAXIMUM: "UtilizationMetricsWriteIOPSEphemeralStorageMaximum",
};
const RDSDBRecommendationFilterName = {
    IDLE: "Idle",
    INSTANCE_FINDING: "InstanceFinding",
    INSTANCE_FINDING_REASON_CODE: "InstanceFindingReasonCode",
    STORAGE_FINDING: "StorageFinding",
    STORAGE_FINDING_REASON_CODE: "StorageFindingReasonCode",
};
const EBSSavingsEstimationModeSource = {
    COST_EXPLORER_RIGHTSIZING: "CostExplorerRightsizing",
    COST_OPTIMIZATION_HUB: "CostOptimizationHub",
    PUBLIC_PRICING: "PublicPricing",
};
const EBSFinding = {
    NOT_OPTIMIZED: "NotOptimized",
    OPTIMIZED: "Optimized",
};
const EBSMetricName = {
    VOLUME_READ_BYTES_PER_SECOND: "VolumeReadBytesPerSecond",
    VOLUME_READ_OPS_PER_SECOND: "VolumeReadOpsPerSecond",
    VOLUME_WRITE_BYTES_PER_SECOND: "VolumeWriteBytesPerSecond",
    VOLUME_WRITE_OPS_PER_SECOND: "VolumeWriteOpsPerSecond",
};
const ExternalMetricStatusCode = {
    DATADOG_INTEGRATION_ERROR: "DATADOG_INTEGRATION_ERROR",
    DYNATRACE_INTEGRATION_ERROR: "DYNATRACE_INTEGRATION_ERROR",
    INSTANA_INTEGRATION_ERROR: "INSTANA_INTEGRATION_ERROR",
    INSUFFICIENT_DATADOG_METRICS: "INSUFFICIENT_DATADOG_METRICS",
    INSUFFICIENT_DYNATRACE_METRICS: "INSUFFICIENT_DYNATRACE_METRICS",
    INSUFFICIENT_INSTANA_METRICS: "INSUFFICIENT_INSTANA_METRICS",
    INSUFFICIENT_NEWRELIC_METRICS: "INSUFFICIENT_NEWRELIC_METRICS",
    INTEGRATION_SUCCESS: "INTEGRATION_SUCCESS",
    NEWRELIC_INTEGRATION_ERROR: "NEWRELIC_INTEGRATION_ERROR",
    NO_EXTERNAL_METRIC_SET: "NO_EXTERNAL_METRIC_SET",
};
const InstanceRecommendationFindingReasonCode = {
    CPU_OVER_PROVISIONED: "CPUOverprovisioned",
    CPU_UNDER_PROVISIONED: "CPUUnderprovisioned",
    DISK_IOPS_OVER_PROVISIONED: "DiskIOPSOverprovisioned",
    DISK_IOPS_UNDER_PROVISIONED: "DiskIOPSUnderprovisioned",
    DISK_THROUGHPUT_OVER_PROVISIONED: "DiskThroughputOverprovisioned",
    DISK_THROUGHPUT_UNDER_PROVISIONED: "DiskThroughputUnderprovisioned",
    EBS_IOPS_OVER_PROVISIONED: "EBSIOPSOverprovisioned",
    EBS_IOPS_UNDER_PROVISIONED: "EBSIOPSUnderprovisioned",
    EBS_THROUGHPUT_OVER_PROVISIONED: "EBSThroughputOverprovisioned",
    EBS_THROUGHPUT_UNDER_PROVISIONED: "EBSThroughputUnderprovisioned",
    GPU_MEMORY_OVER_PROVISIONED: "GPUMemoryOverprovisioned",
    GPU_MEMORY_UNDER_PROVISIONED: "GPUMemoryUnderprovisioned",
    GPU_OVER_PROVISIONED: "GPUOverprovisioned",
    GPU_UNDER_PROVISIONED: "GPUUnderprovisioned",
    MEMORY_OVER_PROVISIONED: "MemoryOverprovisioned",
    MEMORY_UNDER_PROVISIONED: "MemoryUnderprovisioned",
    NETWORK_BANDWIDTH_OVER_PROVISIONED: "NetworkBandwidthOverprovisioned",
    NETWORK_BANDWIDTH_UNDER_PROVISIONED: "NetworkBandwidthUnderprovisioned",
    NETWORK_PPS_OVER_PROVISIONED: "NetworkPPSOverprovisioned",
    NETWORK_PPS_UNDER_PROVISIONED: "NetworkPPSUnderprovisioned",
};
const InstanceIdle = {
    FALSE: "False",
    TRUE: "True",
};
const InstanceState = {
    PENDING: "pending",
    RUNNING: "running",
    SHUTTING_DOWN: "shutting-down",
    STOPPED: "stopped",
    STOPPING: "stopping",
    TERMINATED: "terminated",
};
const PlatformDifference = {
    ARCHITECTURE: "Architecture",
    HYPERVISOR: "Hypervisor",
    INSTANCE_STORE_AVAILABILITY: "InstanceStoreAvailability",
    NETWORK_INTERFACE: "NetworkInterface",
    STORAGE_INTERFACE: "StorageInterface",
    VIRTUALIZATION_TYPE: "VirtualizationType",
};
const RecommendationSourceType = {
    AURORA_DB_CLUSTER_STORAGE: "AuroraDBClusterStorage",
    AUTO_SCALING_GROUP: "AutoScalingGroup",
    EBS_VOLUME: "EbsVolume",
    EC2_INSTANCE: "Ec2Instance",
    ECS_SERVICE: "EcsService",
    LAMBDA_FUNCTION: "LambdaFunction",
    LICENSE: "License",
    NAT_GATEWAY: "NatGateway",
    RDS_DB_INSTANCE: "RdsDBInstance",
    RDS_DB_INSTANCE_STORAGE: "RdsDBInstanceStorage",
};
const ECSServiceMetricName = {
    CPU: "Cpu",
    MEMORY: "Memory",
};
const ECSSavingsEstimationModeSource = {
    COST_EXPLORER_RIGHTSIZING: "CostExplorerRightsizing",
    COST_OPTIMIZATION_HUB: "CostOptimizationHub",
    PUBLIC_PRICING: "PublicPricing",
};
const ECSServiceRecommendationFinding = {
    OPTIMIZED: "Optimized",
    OVER_PROVISIONED: "Overprovisioned",
    UNDER_PROVISIONED: "Underprovisioned",
};
const ECSServiceRecommendationFindingReasonCode = {
    CPU_OVER_PROVISIONED: "CPUOverprovisioned",
    CPU_UNDER_PROVISIONED: "CPUUnderprovisioned",
    MEMORY_OVER_PROVISIONED: "MemoryOverprovisioned",
    MEMORY_UNDER_PROVISIONED: "MemoryUnderprovisioned",
};
const ECSServiceLaunchType = {
    EC2: "EC2",
    FARGATE: "Fargate",
};
const ECSServiceMetricStatistic = {
    AVERAGE: "Average",
    MAXIMUM: "Maximum",
};
const EnrollmentFilterName = {
    STATUS: "Status",
};
const Dimension = {
    SAVINGS_VALUE: "SavingsValue",
    SAVINGS_VALUE_AFTER_DISCOUNT: "SavingsValueAfterDiscount",
};
const Order = {
    ASC: "Asc",
    DESC: "Desc",
};
const IdleRecommendationResourceType = {
    AUTO_SCALING_GROUP: "AutoScalingGroup",
    EBS_VOLUME: "EBSVolume",
    EC2_INSTANCE: "EC2Instance",
    ECS_SERVICE: "ECSService",
    NAT_GATEWAY: "NatGateway",
    RDS_DB_INSTANCE: "RDSDBInstance",
};
const IdleFinding = {
    IDLE: "Idle",
    UNATTACHED: "Unattached",
    UNUSED: "Unused",
};
const IdleMetricName = {
    ACTIVE_CONNECTION_COUNT: "ActiveConnectionCount",
    CPU: "CPU",
    DATABASE_CONNECTIONS: "DatabaseConnections",
    EBS_VOLUME_READ_IOPS: "EBSVolumeReadIOPS",
    EBS_VOLUME_WRITE_IOPS: "EBSVolumeWriteIOPS",
    MEMORY: "Memory",
    NETWORK_IN_BYTES_PER_SECOND: "NetworkInBytesPerSecond",
    NETWORK_OUT_BYTES_PER_SECOND: "NetworkOutBytesPerSecond",
    PACKETS_IN_FROM_DESTINATION: "PacketsInFromDestination",
    PACKETS_IN_FROM_SOURCE: "PacketsInFromSource",
    VOLUME_READ_OPS_PER_SECOND: "VolumeReadOpsPerSecond",
    VOLUME_WRITE_OPS_PER_SECOND: "VolumeWriteOpsPerSecond",
};
const LambdaSavingsEstimationModeSource = {
    COST_EXPLORER_RIGHTSIZING: "CostExplorerRightsizing",
    COST_OPTIMIZATION_HUB: "CostOptimizationHub",
    PUBLIC_PRICING: "PublicPricing",
};
const LambdaFunctionRecommendationFinding = {
    NOT_OPTIMIZED: "NotOptimized",
    OPTIMIZED: "Optimized",
    UNAVAILABLE: "Unavailable",
};
const LambdaFunctionRecommendationFindingReasonCode = {
    INCONCLUSIVE: "Inconclusive",
    INSUFFICIENT_DATA: "InsufficientData",
    MEMORY_OVER_PROVISIONED: "MemoryOverprovisioned",
    MEMORY_UNDER_PROVISIONED: "MemoryUnderprovisioned",
};
const LambdaFunctionMemoryMetricName = {
    DURATION: "Duration",
};
const LambdaFunctionMemoryMetricStatistic = {
    EXPECTED: "Expected",
    LOWER_BOUND: "LowerBound",
    UPPER_BOUND: "UpperBound",
};
const LambdaFunctionMetricName = {
    DURATION: "Duration",
    MEMORY: "Memory",
};
const LambdaFunctionMetricStatistic = {
    AVERAGE: "Average",
    MAXIMUM: "Maximum",
};
const LicenseEdition = {
    ENTERPRISE: "Enterprise",
    FREE: "Free",
    NO_LICENSE_EDITION_FOUND: "NoLicenseEditionFound",
    STANDARD: "Standard",
};
const LicenseModel = {
    BRING_YOUR_OWN_LICENSE: "BringYourOwnLicense",
    LICENSE_INCLUDED: "LicenseIncluded",
};
const LicenseName = {
    SQLSERVER: "SQLServer",
};
const MetricSourceProvider = {
    CloudWatchAppInsights: "CloudWatchApplicationInsights",
};
const LicenseFinding = {
    INSUFFICIENT_METRICS: "InsufficientMetrics",
    NOT_OPTIMIZED: "NotOptimized",
    OPTIMIZED: "Optimized",
};
const LicenseFindingReasonCode = {
    CW_APP_INSIGHTS_DISABLED: "InvalidCloudWatchApplicationInsightsSetup",
    CW_APP_INSIGHTS_ERROR: "CloudWatchApplicationInsightsError",
    LICENSE_OVER_PROVISIONED: "LicenseOverprovisioned",
    OPTIMIZED: "Optimized",
};
const RDSDBMetricName = {
    AURORA_MEMORY_HEALTH_STATE: "AuroraMemoryHealthState",
    AURORA_MEMORY_NUM_DECLINED_SQL: "AuroraMemoryNumDeclinedSql",
    AURORA_MEMORY_NUM_KILL_CONN_TOTAL: "AuroraMemoryNumKillConnTotal",
    AURORA_MEMORY_NUM_KILL_QUERY_TOTAL: "AuroraMemoryNumKillQueryTotal",
    CPU: "CPU",
    DATABASE_CONNECTIONS: "DatabaseConnections",
    EBS_VOLUME_READ_IOPS: "EBSVolumeReadIOPS",
    EBS_VOLUME_READ_THROUGHPUT: "EBSVolumeReadThroughput",
    EBS_VOLUME_STORAGE_SPACE_UTILIZATION: "EBSVolumeStorageSpaceUtilization",
    EBS_VOLUME_WRITE_IOPS: "EBSVolumeWriteIOPS",
    EBS_VOLUME_WRITE_THROUGHPUT: "EBSVolumeWriteThroughput",
    MEMORY: "Memory",
    NETWORK_RECEIVE_THROUGHPUT: "NetworkReceiveThroughput",
    NETWORK_TRANSMIT_THROUGHPUT: "NetworkTransmitThroughput",
    READ_IOPS_EPHEMERAL_STORAGE: "ReadIOPSEphemeralStorage",
    STORAGE_NETWORK_RECEIVE_THROUGHPUT: "StorageNetworkReceiveThroughput",
    STORAGE_NETWORK_TRANSMIT_THROUGHPUT: "StorageNetworkTransmitThroughput",
    VOLUME_BYTES_USED: "VolumeBytesUsed",
    VOLUME_READ_IOPS: "VolumeReadIOPs",
    VOLUME_WRITE_IOPS: "VolumeWriteIOPs",
    WRITE_IOPS_EPHEMERAL_STORAGE: "WriteIOPSEphemeralStorage",
};
const RDSCurrentInstancePerformanceRisk = {
    HIGH: "High",
    LOW: "Low",
    MEDIUM: "Medium",
    VERY_LOW: "VeryLow",
};
const RDSEstimatedMonthlyVolumeIOPsCostVariation = {
    HIGH: "High",
    LOW: "Low",
    MEDIUM: "Medium",
    NONE: "None",
};
const RDSSavingsEstimationModeSource = {
    COST_EXPLORER_RIGHTSIZING: "CostExplorerRightsizing",
    COST_OPTIMIZATION_HUB: "CostOptimizationHub",
    PUBLIC_PRICING: "PublicPricing",
};
const Idle = {
    FALSE: "False",
    TRUE: "True",
};
const RDSInstanceFinding = {
    OPTIMIZED: "Optimized",
    OVER_PROVISIONED: "Overprovisioned",
    UNDER_PROVISIONED: "Underprovisioned",
};
const RDSInstanceFindingReasonCode = {
    CPU_OVER_PROVISIONED: "CPUOverprovisioned",
    CPU_UNDER_PROVISIONED: "CPUUnderprovisioned",
    DB_CLUSTER_WRITER_UNDER_PROVISIONED: "DBClusterWriterUnderprovisioned",
    EBS_IOPS_OVER_PROVISIONED: "EBSIOPSOverprovisioned",
    EBS_IOPS_UNDER_PROVISIONED: "EBSIOPSUnderprovisioned",
    EBS_THROUGHPUT_OVER_PROVISIONED: "EBSThroughputOverprovisioned",
    EBS_THROUGHPUT_UNDER_PROVISIONED: "EBSThroughputUnderprovisioned",
    INSTANCE_STORAGE_READ_IOPS_UNDER_PROVISIONED: "InstanceStorageReadIOPSUnderprovisioned",
    INSTANCE_STORAGE_WRITE_IOPS_UNDER_PROVISIONED: "InstanceStorageWriteIOPSUnderprovisioned",
    MEMORY_UNDER_PROVISIONED: "MemoryUnderprovisioned",
    NETWORK_BANDWIDTH_OVER_PROVISIONED: "NetworkBandwidthOverprovisioned",
    NETWORK_BANDWIDTH_UNDER_PROVISIONED: "NetworkBandwidthUnderprovisioned",
    NEW_ENGINE_VERSION_AVAILABLE: "NewEngineVersionAvailable",
    NEW_GENERATION_DB_INSTANCE_CLASS_AVAILABLE: "NewGenerationDBInstanceClassAvailable",
};
const RDSDBMetricStatistic = {
    AVERAGE: "Average",
    MAXIMUM: "Maximum",
    MINIMUM: "Minimum",
};
const RDSStorageFinding = {
    NOT_OPTIMIZED: "NotOptimized",
    OPTIMIZED: "Optimized",
    OVER_PROVISIONED: "Overprovisioned",
    UNDER_PROVISIONED: "Underprovisioned",
};
const RDSStorageFindingReasonCode = {
    DB_CLUSTER_STORAGE_OPTION_AVAILABLE: "DBClusterStorageOptionAvailable",
    DB_CLUSTER_STORAGE_SAVINGS_AVAILABLE: "DBClusterStorageSavingsAvailable",
    EBS_VOLUME_ALLOCATED_STORAGE_UNDER_PROVISIONED: "EBSVolumeAllocatedStorageUnderprovisioned",
    EBS_VOLUME_IOPS_OVER_PROVISIONED: "EBSVolumeIOPSOverprovisioned",
    EBS_VOLUME_THROUGHPUT_OVER_PROVISIONED: "EBSVolumeThroughputOverprovisioned",
    EBS_VOLUME_THROUGHPUT_UNDER_PROVISIONED: "EBSVolumeThroughputUnderprovisioned",
    NEW_GENERATION_STORAGE_TYPE_AVAILABLE: "NewGenerationStorageTypeAvailable",
};
const SavingsEstimationMode = {
    AFTER_DISCOUNTS: "AfterDiscounts",
    BEFORE_DISCOUNTS: "BeforeDiscounts",
};
const FindingReasonCode = {
    MEMORY_OVER_PROVISIONED: "MemoryOverprovisioned",
    MEMORY_UNDER_PROVISIONED: "MemoryUnderprovisioned",
};

exports.$Command = smithyClient.Command;
exports.__Client = smithyClient.Client;
exports.ComputeOptimizerServiceException = ComputeOptimizerServiceException.ComputeOptimizerServiceException;
exports.AllocationStrategy = AllocationStrategy;
exports.AsgType = AsgType;
exports.AutoScalingConfiguration = AutoScalingConfiguration;
exports.ComputeOptimizer = ComputeOptimizer;
exports.ComputeOptimizerClient = ComputeOptimizerClient;
exports.CpuVendorArchitecture = CpuVendorArchitecture;
exports.Currency = Currency;
exports.CurrentPerformanceRisk = CurrentPerformanceRisk;
exports.CustomizableMetricHeadroom = CustomizableMetricHeadroom;
exports.CustomizableMetricName = CustomizableMetricName;
exports.CustomizableMetricThreshold = CustomizableMetricThreshold;
exports.DeleteRecommendationPreferencesCommand = DeleteRecommendationPreferencesCommand;
exports.DescribeRecommendationExportJobsCommand = DescribeRecommendationExportJobsCommand;
exports.Dimension = Dimension;
exports.EBSFilterName = EBSFilterName;
exports.EBSFinding = EBSFinding;
exports.EBSMetricName = EBSMetricName;
exports.EBSSavingsEstimationModeSource = EBSSavingsEstimationModeSource;
exports.ECSSavingsEstimationModeSource = ECSSavingsEstimationModeSource;
exports.ECSServiceLaunchType = ECSServiceLaunchType;
exports.ECSServiceMetricName = ECSServiceMetricName;
exports.ECSServiceMetricStatistic = ECSServiceMetricStatistic;
exports.ECSServiceRecommendationFilterName = ECSServiceRecommendationFilterName;
exports.ECSServiceRecommendationFinding = ECSServiceRecommendationFinding;
exports.ECSServiceRecommendationFindingReasonCode = ECSServiceRecommendationFindingReasonCode;
exports.EnhancedInfrastructureMetrics = EnhancedInfrastructureMetrics;
exports.EnrollmentFilterName = EnrollmentFilterName;
exports.ExportAutoScalingGroupRecommendationsCommand = ExportAutoScalingGroupRecommendationsCommand;
exports.ExportEBSVolumeRecommendationsCommand = ExportEBSVolumeRecommendationsCommand;
exports.ExportEC2InstanceRecommendationsCommand = ExportEC2InstanceRecommendationsCommand;
exports.ExportECSServiceRecommendationsCommand = ExportECSServiceRecommendationsCommand;
exports.ExportIdleRecommendationsCommand = ExportIdleRecommendationsCommand;
exports.ExportLambdaFunctionRecommendationsCommand = ExportLambdaFunctionRecommendationsCommand;
exports.ExportLicenseRecommendationsCommand = ExportLicenseRecommendationsCommand;
exports.ExportRDSDatabaseRecommendationsCommand = ExportRDSDatabaseRecommendationsCommand;
exports.ExportableAutoScalingGroupField = ExportableAutoScalingGroupField;
exports.ExportableECSServiceField = ExportableECSServiceField;
exports.ExportableIdleField = ExportableIdleField;
exports.ExportableInstanceField = ExportableInstanceField;
exports.ExportableLambdaFunctionField = ExportableLambdaFunctionField;
exports.ExportableLicenseField = ExportableLicenseField;
exports.ExportableRDSDBField = ExportableRDSDBField;
exports.ExportableVolumeField = ExportableVolumeField;
exports.ExternalMetricStatusCode = ExternalMetricStatusCode;
exports.ExternalMetricsSource = ExternalMetricsSource;
exports.FileFormat = FileFormat;
exports.FilterName = FilterName;
exports.Finding = Finding;
exports.FindingReasonCode = FindingReasonCode;
exports.GetAutoScalingGroupRecommendationsCommand = GetAutoScalingGroupRecommendationsCommand;
exports.GetEBSVolumeRecommendationsCommand = GetEBSVolumeRecommendationsCommand;
exports.GetEC2InstanceRecommendationsCommand = GetEC2InstanceRecommendationsCommand;
exports.GetEC2RecommendationProjectedMetricsCommand = GetEC2RecommendationProjectedMetricsCommand;
exports.GetECSServiceRecommendationProjectedMetricsCommand = GetECSServiceRecommendationProjectedMetricsCommand;
exports.GetECSServiceRecommendationsCommand = GetECSServiceRecommendationsCommand;
exports.GetEffectiveRecommendationPreferencesCommand = GetEffectiveRecommendationPreferencesCommand;
exports.GetEnrollmentStatusCommand = GetEnrollmentStatusCommand;
exports.GetEnrollmentStatusesForOrganizationCommand = GetEnrollmentStatusesForOrganizationCommand;
exports.GetIdleRecommendationsCommand = GetIdleRecommendationsCommand;
exports.GetLambdaFunctionRecommendationsCommand = GetLambdaFunctionRecommendationsCommand;
exports.GetLicenseRecommendationsCommand = GetLicenseRecommendationsCommand;
exports.GetRDSDatabaseRecommendationProjectedMetricsCommand = GetRDSDatabaseRecommendationProjectedMetricsCommand;
exports.GetRDSDatabaseRecommendationsCommand = GetRDSDatabaseRecommendationsCommand;
exports.GetRecommendationPreferencesCommand = GetRecommendationPreferencesCommand;
exports.GetRecommendationSummariesCommand = GetRecommendationSummariesCommand;
exports.Idle = Idle;
exports.IdleFinding = IdleFinding;
exports.IdleMetricName = IdleMetricName;
exports.IdleRecommendationFilterName = IdleRecommendationFilterName;
exports.IdleRecommendationResourceType = IdleRecommendationResourceType;
exports.InferredWorkloadType = InferredWorkloadType;
exports.InferredWorkloadTypesPreference = InferredWorkloadTypesPreference;
exports.InstanceIdle = InstanceIdle;
exports.InstanceRecommendationFindingReasonCode = InstanceRecommendationFindingReasonCode;
exports.InstanceSavingsEstimationModeSource = InstanceSavingsEstimationModeSource;
exports.InstanceState = InstanceState;
exports.JobFilterName = JobFilterName;
exports.JobStatus = JobStatus;
exports.LambdaFunctionMemoryMetricName = LambdaFunctionMemoryMetricName;
exports.LambdaFunctionMemoryMetricStatistic = LambdaFunctionMemoryMetricStatistic;
exports.LambdaFunctionMetricName = LambdaFunctionMetricName;
exports.LambdaFunctionMetricStatistic = LambdaFunctionMetricStatistic;
exports.LambdaFunctionRecommendationFilterName = LambdaFunctionRecommendationFilterName;
exports.LambdaFunctionRecommendationFinding = LambdaFunctionRecommendationFinding;
exports.LambdaFunctionRecommendationFindingReasonCode = LambdaFunctionRecommendationFindingReasonCode;
exports.LambdaSavingsEstimationModeSource = LambdaSavingsEstimationModeSource;
exports.LicenseEdition = LicenseEdition;
exports.LicenseFinding = LicenseFinding;
exports.LicenseFindingReasonCode = LicenseFindingReasonCode;
exports.LicenseModel = LicenseModel;
exports.LicenseName = LicenseName;
exports.LicenseRecommendationFilterName = LicenseRecommendationFilterName;
exports.LookBackPeriodPreference = LookBackPeriodPreference;
exports.MetricName = MetricName;
exports.MetricSourceProvider = MetricSourceProvider;
exports.MetricStatistic = MetricStatistic;
exports.MigrationEffort = MigrationEffort;
exports.Order = Order;
exports.PlatformDifference = PlatformDifference;
exports.PreferredResourceName = PreferredResourceName;
exports.PutRecommendationPreferencesCommand = PutRecommendationPreferencesCommand;
exports.RDSCurrentInstancePerformanceRisk = RDSCurrentInstancePerformanceRisk;
exports.RDSDBMetricName = RDSDBMetricName;
exports.RDSDBMetricStatistic = RDSDBMetricStatistic;
exports.RDSDBRecommendationFilterName = RDSDBRecommendationFilterName;
exports.RDSEstimatedMonthlyVolumeIOPsCostVariation = RDSEstimatedMonthlyVolumeIOPsCostVariation;
exports.RDSInstanceFinding = RDSInstanceFinding;
exports.RDSInstanceFindingReasonCode = RDSInstanceFindingReasonCode;
exports.RDSSavingsEstimationModeSource = RDSSavingsEstimationModeSource;
exports.RDSStorageFinding = RDSStorageFinding;
exports.RDSStorageFindingReasonCode = RDSStorageFindingReasonCode;
exports.RecommendationPreferenceName = RecommendationPreferenceName;
exports.RecommendationSourceType = RecommendationSourceType;
exports.ResourceType = ResourceType;
exports.SavingsEstimationMode = SavingsEstimationMode;
exports.ScopeName = ScopeName;
exports.Status = Status;
exports.UpdateEnrollmentStatusCommand = UpdateEnrollmentStatusCommand;
exports.paginateDescribeRecommendationExportJobs = paginateDescribeRecommendationExportJobs;
exports.paginateGetEnrollmentStatusesForOrganization = paginateGetEnrollmentStatusesForOrganization;
exports.paginateGetLambdaFunctionRecommendations = paginateGetLambdaFunctionRecommendations;
exports.paginateGetRecommendationPreferences = paginateGetRecommendationPreferences;
exports.paginateGetRecommendationSummaries = paginateGetRecommendationSummaries;
Object.prototype.hasOwnProperty.call(schemas_0, '__proto__') &&
    !Object.prototype.hasOwnProperty.call(exports, '__proto__') &&
    Object.defineProperty(exports, '__proto__', {
        enumerable: true,
        value: schemas_0['__proto__']
    });

Object.keys(schemas_0).forEach(function (k) {
    if (k !== 'default' && !Object.prototype.hasOwnProperty.call(exports, k)) exports[k] = schemas_0[k];
});
Object.prototype.hasOwnProperty.call(errors, '__proto__') &&
    !Object.prototype.hasOwnProperty.call(exports, '__proto__') &&
    Object.defineProperty(exports, '__proto__', {
        enumerable: true,
        value: errors['__proto__']
    });

Object.keys(errors).forEach(function (k) {
    if (k !== 'default' && !Object.prototype.hasOwnProperty.call(exports, k)) exports[k] = errors[k];
});
